<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Coalescent Analyses</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
        <li><class="header-search">
  <form class="header-search-form" action="/search.html" method="get">
    <input type="text" id="search-box-nav" placeholder="Search..." name="query" style="border-color:black;background-color:white;height:32px;">
    <button type="submit" style="color:black; border-color:black;width:32px;height:32px;padding-top:4px">
        <img src="https://revbayes.github.io/assets/img/search.png" height="22px" width="28px"/>
    </button>
  </form>
</div>

</li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <!-- adds margin to main content -->
      <div class="container" style="padding-left:20px padding-right:20px;">
          <div class="titlebar">
	<h1 class="maintitle">Coalescent Analyses</h1>
	<h3 class="subtitle">Estimating Demographic Histories with Bayesian Coalescent Skyline Plot Models</h3>
	<h4 class="authors">Ronja Billenstein and Sebastian Höhna</h4>
  <h5>Last modified on June 11, 2024</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/intro/">Getting Started with RevBayes and Rev Language Syntax</a></li>
          
            <li><a href="/tutorials/mcmc/">Introduction to Markov chain Monte Carlo (MCMC) Sampling</a></li>
          
            <li><a href="/tutorials/ctmc/">Nucleotide substitution models</a></li>
          
            <li><a href="/tutorials/mcmc/moves.html">Overview of Moves in MCMC</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Other</strong>
        <ul id="other_files">
        
        
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_ages.tsv">horses_heterochronous_ages.tsv</a></li>
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_sequences.fasta">horses_heterochronous_sequences.fasta</a></li>
        
          <li><a href="/tutorials/coalescent/data/horses_isochronous_sequences.fasta">horses_isochronous_sequences.fasta</a></li>
        
        </ul>
    
</blockquote>


</div>
<h2 class="section" id="overview">Overview</h2>
<hr class="section" />

<p>This tutorial describes how to run a demographic analysis using the coalescent process in <code class="language-plaintext highlighter-rouge">RevBayes</code>.
Demographic inference is about estimating population dynamics and in this tutorial, we will specifically focus on population size estimation.
The coalescent process provides a flexible way of estimating population size trajectories through time.
We are primarily interested in asking:
(1) what the population size for our given population was?
and (2) how population sizes have changed over time, <em>e.g.</em>, increased or decreased towards the present or experienced a bottleneck.
The input data are usually sequence alignments or estimated trees.
In case of sequence data, trees and population size parameters will be jointly estimated.
Here, we consider different types of analysis, starting with a constant demographic history and doing more complex analyses in later exercises.
The first analyses will consider data from isochronous samples, <em>i.e.</em>, data from samples that have all been collected at the same time.
In the second part of the exercises, you will be asked to also analyze data from heterochronous samples, <em>i.e.</em>, samples that have been collected at different points in time.
The data for these analyes are taken from <a class="citation" href="#Vershinina2021">Vershinina et al. (2021)</a>.</p>

<p><br />
<br />
<br /></p>

<p>The following table provides an overview of different coalescent models. The tutorials column links directly to the respective tutorial with isochronous data. The tutorial on <a href="/tutorials/coalescent/heterochronous">heterochronous data</a> shows how to apply the models to data from different points in time.</p>

<table>
  <thead>
    <tr>
      <th>Model</th>
      <th>Description</th>
      <th>Demographic function within intervals</th>
      <th>Prior on the population sizes</th>
      <th>Interval change-point method</th>
      <th>Multiple loci analysis possible?</th>
      <th>Tutorial</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Constant</td>
      <td>A simple coalescent model with one constant population size.</td>
      <td>constant</td>
      <td>individually chosen</td>
      <td>-</td>
      <td>Yes</td>
      <td><a href="/tutorials/coalescent/Constant">Constant</a></td>
    </tr>
    <tr>
      <td>Linear</td>
      <td>A coalescent model with a linear population size trajectory.</td>
      <td>linear</td>
      <td>individually chosen</td>
      <td>-</td>
      <td>Yes</td>
      <td> </td>
    </tr>
    <tr>
      <td>Exponential</td>
      <td>A coalescent model with an exponential population size trajectory.</td>
      <td>exponential</td>
      <td>individually chosen</td>
      <td>-</td>
      <td>Yes</td>
      <td> </td>
    </tr>
    <tr>
      <td>Bayesian Skyline Plot without autocorrelation</td>
      <td>A coalescent model with a piecewise constant population size trajectory. The different pieces (intervals) are not correlated. The times of interval change are dependent on coalescent events (“coalescent event based”).</td>
      <td>piecewise constant</td>
      <td>individually chosen, independent and identically distributed (iid)</td>
      <td>coalescent event based</td>
      <td>No, only single loci</td>
      <td><a href="/tutorials/coalescent/Skyline">Skyline</a></td>
    </tr>
    <tr>
      <td>Bayesian Skyline Plot with autocorrelation<br /><a class="citation" href="#Drummond2005">(Drummond et al. 2005)</a></td>
      <td>A coalescent model with a piecewise constant population size trajectory. The different pieces (intervals) are correlated, each population size is drawn from an exponential distribution with the mean being the previous population size. The times of interval change are dependent on coalescent events (“coalescent event based”).</td>
      <td>piecewise constant</td>
      <td>$\theta_i | \theta_{i-1}$ (exponential distribution with mean $\theta_{i-1}$)</td>
      <td>coalescent event based</td>
      <td>No, only single loci</td>
      <td>Little description included in <a href="/tutorials/coalescent/Skyline#secAltPriors">Skyline</a><br /><br /><a href="scripts/mcmc_isochronous_BSP.Rev">Example script</a></td>
    </tr>
    <tr>
      <td>Extended Bayesian Skyline Plot<br /><a class="citation" href="#Heled2008">(Heled and Drummond 2008)</a></td>
      <td>A coalescent model with a piecewise constant or linear population size trajectory. The different pieces are not correlated. The times of interval change are dependent on coalescent events (“coalescent event based”), they are independent and identically distributed (iid). The number of changes and thus the number of intervals is determined by stochastic variable search.</td>
      <td>piecewise constant or linear</td>
      <td>independent and identically distributed (iid)</td>
      <td>coalescent event based - number and position of change points estimated</td>
      <td>Yes</td>
      <td>Little description included in <a href="/tutorials/coalescent/Skyline#secAltPriors">Skyline</a><br /><br /><a href="scripts/mcmc_isochronous_EBSP.Rev">Example script</a></td>
    </tr>
    <tr>
      <td>Skyride<br /><a class="citation" href="#Minin2008">(Minin et al. 2008)</a></td>
      <td>A coalescent model with a piecewise constant population size trajectory. The different pieces are correlated via a Gaussian Markov Random Field (GMRF) Prior, the degree of smoothing is determined by a precision parameter. The times of interval change are dependent on coalescent events (“coalescent event based”).</td>
      <td>piecewise constant</td>
      <td>$\theta_i | \theta_{i-1}$ (GMRF Prior with precision parameter)</td>
      <td>coalescent event based</td>
      <td>No, only single loci</td>
      <td>Little description included in <a href="/tutorials/coalescent/Skyline#secAltPriors">Skyline</a><br /><br /><a href="scripts/mcmc_isochronous_Skyride.Rev">Example script</a></td>
    </tr>
    <tr>
      <td>Skygrid<br /><a class="citation" href="#Gill2012">(Gill et al. 2012)</a></td>
      <td>A coalescent model with a piecewise constant population size trajectory. The different pieces are correlated via a Gaussian Markov Random Field (GMRF) Prior and are equally spaced in time, the degree of smoothing is determined by a precision parameter. The times of interval change are independent from coalescent events, they are independent and identically distributed (iid).</td>
      <td>piecewise constant</td>
      <td>$\theta_i | \theta_{i-1}$ (GMRF Prior with precision parameter)</td>
      <td>coalescent event independent - equally sized</td>
      <td>Yes</td>
      <td>Little description included in <a href="/tutorials/coalescent/GMRF#secAltPriors">GMRF</a><br /><br /><a href="scripts/mcmc_isochronous_Skygrid.Rev">Example script</a></td>
    </tr>
    <tr>
      <td>Gaussian Markov Random Field (GMRF) Prior with interval change points independent from coalescent events  <a class="citation" href="#Faulkner2020">(Faulkner et al. 2020)</a></td>
      <td>A coalescent model with a piecewise constant or linear population size trajectory. The different pieces are correlated via a Gaussian Markov Random Field (GMRF) Prior. The times of interval change are independent from coalescent events, they are independent and identically distributed (iid).</td>
      <td>piecewise constant or linear</td>
      <td>$\theta_i | \theta_{i-1}$ (GMRF Prior)</td>
      <td>coalescent event independent, user-specified - here equally sized</td>
      <td>Yes</td>
      <td>With sequences as input data: <a href="/tutorials/coalescent/GMRF">GMRF</a><br /><br />With trees as input data: <a href="/tutorials/coalescent/GMRF_treebased">GMRF treebased</a></td>
    </tr>
    <tr>
      <td>Horseshoe Markov Random Field (HSMRF) Prior with interval change points independent from coalescent events <a class="citation" href="#Faulkner2020">(Faulkner et al. 2020)</a></td>
      <td>A coalescent model with a piecewise constant or linear population size trajectory. The different pieces are correlated via a Horseshoe Markov Random Field (HSMRF) Prior. The times of interval change are independent from coalescent events, they are independent and identically distributed (iid).</td>
      <td>piecewise constant or linear</td>
      <td>$\theta_i | \theta_{i-1}$ (HSMRF Prior)</td>
      <td>coalescent event independent, user-specified - here equally sized</td>
      <td>Yes</td>
      <td>Little description included in <a href="/tutorials/coalescent/GMRF#secHSMRF">GMRF</a><br /><br />Script in <a href="/tutorials/coalescent/HSMRF">HSMRF</a></td>
    </tr>
    <tr>
      <td>Skyfish model<br />(similar to <a class="citation" href="#OpgenRhein2005">Opgen-Rhein et al. (2005)</a>)</td>
      <td>A coalescent model with a piecewise constant or linear population size trajectory. The number of pieces (intervals) is not fixed. A Poisson Prior is used on the number of interval change points, additional priors need to be defined for population sizes and change points. The number of change points is determined via reversible jump MCMC (rjMCMC).</td>
      <td>piecewise constant or linear</td>
      <td>independent and identically distributed (iid)</td>
      <td>coalescent event independent, number and position of change points estimated</td>
      <td>Yes</td>
      <td><a href="/tutorials/coalescent/Skyfish">Skyfish</a></td>
    </tr>
    <tr>
      <td>Piecewise<br />(similar to <a class="citation" href="#Pybus2002">Pybus and Rambaut (2002)</a>, but with flexible combinations)</td>
      <td>A coalescent model with user-defined demographic functions for a user-defined number of intervals. Base demographic functions included in RevBayes are constant, linear and exponential population size trajectories. These can be combined in an arbitrary way, but the most ancient one should always be constant due to computational reasons.</td>
      <td>piecewise constant, linear, or exponential; can be different for every interval</td>
      <td>individually chosen, independent and identically distributed (iid)</td>
      <td>coalescent event independent, user-specified</td>
      <td>No, only single loci</td>
      <td><a href="/tutorials/coalescent/piecewise">Piecewise</a></td>
    </tr>
  </tbody>
</table>

<!--- ### Why?! --->

<!---
<h3 class="subsection" id="the-coalescent">The Coalescent</h3><hr class="subsection">
The coalescent process is constructing a tree backwards in time.
Starting from the samples, lineages are merged (*i.e.* coalesced), always two at a time.
Under the coalescent process, the waiting time between two coalescent events is exponentially distributed and depends on the number of 'active' lineages and the effective population size $N_e$.
Active lineages are the ones that can coalesce, the number is reduced by one with every coalescent event.
The coalescent process was first introduced by Kingman in 1982 for a constant population size <a class="citation" href="#Kingman1982">(Kingman 1982)</a>.
Griffiths and Tavaré then extended the model to be applicable to varying population sizes <a class="citation" href="#Griffiths1994">(Griffiths et al. 1994)</a>.
--->
<!--- Also, samples from different ages can be included <a class="citation" href="#"></a>. (Should go to heterochronous part) --->

<!--- ### Add figure!! --->

<h2 class="section" id="preparation">Preparation</h2>
<hr class="section" />

<blockquote class="instruction">
  <p>In order to perform the different analyses in this tutorial, you will need to create a directory on your computer for this tutorial and download a few files.</p>
</blockquote>

<h3 class="subsection" id="the-data">The Data</h3>
<hr class="subsection" />

<blockquote class="instruction">
  <p>In the tutorial directory on your computer, create a subdirectory called “data” and download the data files that you can find on the left of this page.</p>
</blockquote>

<p>You should now have the following files in your data folder:</p>

<ul>
  <li>
    <p>horses_isochronous_sequences.fasta: an alignment in FASTA format, containing sequences of 36 horse taxa, all sampled at the same time.</p>
  </li>
  <li>
    <p>horses_heterochronous_sequences.fasta: an alignment in FASTA format, containing sequences of 173 horse taxa, sampled at various times.</p>
  </li>
  <li>
    <p>horses_heterochronous_ages.tsv: a tab seperated table listing the heterochronous horse samples and their ages. For extant taxa, the minimum age is 0.0 (i.e. the present).</p>
  </li>
</ul>

<h3 class="subsection" id="the-scripts">The Scripts</h3>
<hr class="subsection" />

<p>It is useful to create <code class="language-plaintext highlighter-rouge">.Rev</code> scripts for the different analyes.
We will also provide a full script in every tutorial that you can easily run for the whole analysis.</p>

<blockquote class="instruction">
  <p>Please create a “scripts” directory in the tutorial directory.</p>
</blockquote>

<h3 class="subsection" id="the-figures">The Figures</h3>
<hr class="subsection" />

<p>The <code class="language-plaintext highlighter-rouge">R</code> package <code class="language-plaintext highlighter-rouge">RevGadgets</code> provides functionality for plotting <code class="language-plaintext highlighter-rouge">RevBayes</code> results.
In each exercise, you can plot the population size trajectories resulting from the different analyses.</p>

<blockquote class="instruction">
  <p>If you want to plot the results from the exercises, please create a “figures” directory in the tutorial directory and install the <code class="language-plaintext highlighter-rouge">R</code> package <code class="language-plaintext highlighter-rouge">RevGadgets</code>.</p>
</blockquote>

<h2 class="section" id="exercises">Exercises</h2>
<hr class="section" />

<blockquote class="instruction">
  <p>You can begin by clicking on the first exercise!</p>
</blockquote>

<p>If you want to turn back to this page, it is listed on the left as prerequisite in all the exercises.
The exercise for heterochronous data is based on the exercises for isochronous data.
It aims at highlighting the changes you have to make when considering samples with different ages.
It is therefore recommended to do the exercises for isochronous data first.</p>

<p>The first five exercises work with isochronous data, the last one with heterochronous data:</p>
<ol>
  <li><a href="/tutorials/coalescent/Constant">The Constant coalescent model</a></li>
  <li><a href="/tutorials/coalescent/Skyline">The Skyline model</a></li>
  <li><a href="/tutorials/coalescent/GMRF">The Gaussian Markov Random Field (GMRF) model</a></li>
  <li><a href="/tutorials/coalescent/Skyfish">The Skyfish model</a></li>
  <li><a href="/tutorials/coalescent/GMRF_treebased">The GMRF model with trees as input data</a></li>
  <li><a href="/tutorials/coalescent/piecewise">A piecewise model</a></li>
  <li><a href="/tutorials/coalescent/heterochronous">Heterochronous data</a></li>
</ol>

<h2 class="section" id="summary">Summary</h2>
<hr class="section" />

<p>After doing all the exercises, you can compare the resulting population sizes.
Have a look at the <a href="/tutorials/coalescent/summary">summary</a>.</p>

<ol class="bibliography"><li><span id="Drummond2005">Drummond A.J., Rambaut A., Shapiro B., Pybus O.G. 2005. Bayesian Coalescent Inference of Past Population Dynamics from Molecular Sequences. Molecular Biology and Evolution. 22:1185–1192.</span>

<a href="https://doi.org/10.1093/molbev/msi103">10.1093/molbev/msi103</a>

</li>
<li><span id="Faulkner2020">Faulkner J.R., Magee A.F., Shapiro B., Minin V.N. 2020. Horseshoe-based Bayesian nonparametric estimation of effective population size trajectories. Biometrics. 76:677–690.</span>

<a href="https://doi.org/https://doi.org/10.1111/biom.13276">https://doi.org/10.1111/biom.13276</a>

</li>
<li><span id="Gill2012">Gill M.S., Lemey P., Faria N.R., Rambaut A., Shapiro B., Suchard M.A. 2012. Improving Bayesian Population Dynamics Inference: A Coalescent-Based Model for Multiple Loci. Molecular Biology and Evolution. 30:713–724.</span>

<a href="https://doi.org/10.1093/molbev/mss265">10.1093/molbev/mss265</a>

</li>
<li><span id="Griffiths1994">Griffiths R.C., Tavare S., Bodmer W.F., Donnelly P.J. 1994. Sampling theory for neutral alleles in a varying environment. Philosophical Transactions of the Royal Society of London. Series B: Biological Sciences. 344:403–410.</span>

<a href="https://doi.org/10.1098/rstb.1994.0079">10.1098/rstb.1994.0079</a>

</li>
<li><span id="Heled2008">Heled J., Drummond A.J. 2008. Bayesian inference of population size history from multiple loci. BMC Evolutionary Biology. 8:289.</span>

<a href="https://doi.org/10.1186/1471-2148-8-289">10.1186/1471-2148-8-289</a>

</li>
<li><span id="Kingman1982">Kingman J.F.C. 1982. On the Genealogy of Large Populations. Journal of Applied Probability. 19:27–43.</span>

<a href="https://doi.org/10.2307/3213548">10.2307/3213548</a>

</li>
<li><span id="Minin2008">Minin V.N., Bloomquist E.W., Suchard M.A. 2008. Smooth Skyride through a Rough Skyline: Bayesian Coalescent-Based Inference of Population Dynamics. Molecular Biology and Evolution. 25:1459–1471.</span>

<a href="https://doi.org/10.1093/molbev/msn090">10.1093/molbev/msn090</a>

</li>
<li><span id="OpgenRhein2005">Opgen-Rhein R., Fahrmeir L., Strimmer K. 2005. Inference of demographic history from genealogical trees using reversible jump Markov chain Monte Carlo. BMC Evolutionary Biology. 5:6.</span>

<a href="https://doi.org/10.1186/1471-2148-5-6">10.1186/1471-2148-5-6</a>

</li>
<li><span id="Pybus2002">Pybus O.G., Rambaut A. 2002. GENIE: estimating demographic history from molecular phylogenies. Bioinformatics. 18:1404–1405.</span>

<a href="https://doi.org/10.1093/bioinformatics/18.10.1404">10.1093/bioinformatics/18.10.1404</a>

</li>
<li><span id="Vershinina2021">Vershinina A.O., Heintzman P.D., Froese D.G., Zazula G., Cassatt-Johnstone M., Dalén L., Der Sarkissian C., Dunn S.G., Ermini L., Gamba C., Groves P., Kapp J.D., Mann D.H., Seguin-Orlando A., Southon J., Stiller M., Wooller M.J., Baryshnikov G., Gimranov D., Scott E., Hall E., Hewitson S., Kirillova I., Kosintsev P., Shidlovsky F., Tong H.-W., Tiunov M.P., Vartanyan S., Orlando L., Corbett-Detig R., MacPhee R.D., Shapiro B. 2021. Ancient horse genomes reveal the timing and extent of dispersals across the Bering Land Bridge. Molecular Ecology. 30:6144–6161.</span>

<a href="https://doi.org/https://doi.org/10.1111/mec.15977">https://doi.org/10.1111/mec.15977</a>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      </div>
      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
