<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Background on state-dependent diversification rate estimation</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Background on state-dependent diversification rate estimation</h1>
	<h3 class="subtitle">An introduction to inference using state-dependent speciation and extinction (SSE) branching processes</h3>
	<h4 class="authors">Sebastian Höhna, Will Freyman, and Emma Goldberg</h4>
  <h5>Last modified on March 11, 2022</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul>
          <li>None</li>
          </ul>
        
    </div>
  </div>
  
</blockquote>





</div>
<h2 class="section" id="introduction">Introduction</h2>
<hr class="section" />

<p>This is a general introduction to character state-dependent
branching process models, particularly as they are implemented in RevBayes. Frequently referred to as
state-dependent speciation and extinction (SSE) models, these models are
a birth-death process where the diversification rates are dependent on
the state of an evolving character. The original model of this type
considered a binary character (a trait with two discrete state values;
called BiSSE, <a class="citation" href="#Maddison2007">(Maddison et al. 2007)</a>. Several variants have also been developed
for other types of traits
<a class="citation" href="#FitzJohn2010">(FitzJohn 2010; Goldberg et al. 2011; Goldberg and Igić 2012; Magnuson-Ford and Otto 2012; FitzJohn 2012; Beaulieu and O’Meara 2016; Freyman and Höhna 2018)</a>.</p>

<p>RevBayes can be used to specify a wide range of SSE models. For specific examples see these other tutorials:</p>
<ul>
  <li>BiSSE and MuSSE models: <a href="/tutorials/sse/bisse.html">State-dependent diversification with BiSSE and MuSSE</a></li>
  <li>HiSSE models: <a href="/tutorials/sse/hisse.html">State-dependent diversification with HiSSE</a></li>
  <li>ClaSSE models: <a href="/tutorials/sse/classe.html">State-dependent diversification with the ClaSSE model</a></li>
  <li>ChromoSSE: <a href="/tutorials/chromo/">Chromosome Evolution</a></li>
</ul>

<h2 class="section" id="BiSSE_background">Background: The BiSSE Model</h2>
<hr class="section" />

<p>The binary state speciation and extinction model (BiSSE) <a class="citation" href="#Maddison2007">(Maddison et al. 2007)</a>
was introduced because of two problems identified by <a class="citation" href="#Maddison2006b">Maddison (2006)</a>.
First, inferences about character state transitions based on simple
transition models [like <a class="citation" href="#Pagel1999">Pagel (1999)</a>] can be thrown off if the character
affects rates of speciation or extinction. Second, inferences about
whether a character affects lineage diversification based on sister
clade comparisons <a class="citation" href="#Mitter1988">(Mitter et al. 1988)</a> can be thrown off if the transition
rates are asymmetric. BiSSE and related models are now mostly used to
assess if the states of a character are associated with different rates
of speciation or extinction.</p>

<p>RevBayes implements the extension of BiSSE to any number of discrete
states–<em>i.e.</em>, the MuSSE model in <a href="http://www.zoology.ubc.ca/prog/diversitree/">diversitree</a>;
<a class="citation" href="#FitzJohn2012">(FitzJohn 2012)</a>. We will first describe the general theory about the
model.</p>

<h2 class="section" id="bisse_theory">The theory behind state-dependent diversification models</h2>
<hr class="section" />

<figure id="model_fig"><p><img src="figures/BiSSE.png" width="75%" /></p>
<figcaption>A schematic overview of the BiSSE model.
Each lineage has a binary trait associated with it, so it is either in state 0 (blue) or state 1 (red).
When a lineage is in state 0, it can either (a) speciate with rate $\lambda_0$ which results into two descendant lineage both being in state 0; (b) go extinct with rate $\mu_0$; or (c) transition to state 1 with rate $q_{01}$.
The same types of events are possible when a lineage is in state 1 but with rates $\lambda_1$, $\mu_1$, and $q_{10}$, respectively.</figcaption>
</figure>

<h3 class="subsection" id="general_approach">General approach</h3>
<hr class="subsection" />

<p>The BiSSE model assumes two discrete states
(<em>i.e.,</em> a binary character), and that the
state of each extant species is known
(<em>i.e.,</em> the discrete-valued character is
observed). The general approach adopted by BiSSE and
related models is to derive a set of ordinary differential equations
(ODEs) that describe how the probability of observing a descendant clade
changes along a branch in the observed phylogeny. Each equation in this
set describes how the probability of observing a clade changes through
time if it is in a particular state over that time period; collectively,
these equations are called $\frac{\mathrm{d}D_{N,i}(t)}{\mathrm{d}t}$,
where $i$ is the state of a lineage at time $t$ and $N$ is the clade
descended from that lineage.</p>

<p>Computing the likelihood proceeds by establishing an initial value
problem. We initialize the procedure by observing the character states
of some lineages, generally the tip states. Then starting from those
probabilities (<em>e.g.,</em> species X has state 0
with probability 1 at the present), we describe how those probabilities
change over time (described by the ODEs), working our way back until we
have computed the probabilities of observing that collection of lineages
at some earlier time (<em>e.g.,</em> the root).</p>

<p>As we integrate from the tips to the root, we need to deal with branches
coming together at nodes. Assuming that the parent and daughter lineages
have the same state, we multiply together the probabilities that the
daughters are state $i$ and the instantaneous speciation rate
$\lambda_i$ to get the initial value for the ancestral branch subtending
that node.</p>

<p>Proceeding in this way down the tree results in a set of $k$
probabilities at the root; these $k$ probabilities represent the
probability of observing the phylogeny conditional on the root being in
each of the states (<em>i.e.,</em> the $i^\text{th}$
conditional probability is the probability of observing the tree given
that the root is in state $i$). The overall likelihood of the tree is a
weighted average of the $k$ probabilities at the root, where the
weighting scheme represents the assumed probability that the root was in
each of the $k$ states.</p>

<p>As with all birth-death process models, special care must be taken to
account for the possibility of extinction. Specifically, the above ODEs
must accommodate lineages that may arise along each branch in the tree
that subsequently go extinct before the present (and so are unobserved).
This requires a second set of $k$ ODEs,
$\frac{ \mathrm{d}E_{i}(t)}{\mathrm{d}t}$, which define how the
probability of eventual extinction from state $i$ changes over time.
These ODEs must be solved to compute the differential equations
$\frac{ \mathrm{d}D_{N,i}(t)}{\mathrm{d}t}$. We will derive both sets of
equations in the following sections.</p>

<h3 class="subsection" id="derivation-for-the-binary-state-birth-death-process">Derivation for the binary state birth-death process</h3>
<hr class="subsection" />

<p>The derivation here follows the original description in <a class="citation" href="#Maddison2007">Maddison et al. (2007)</a>.
Consider a (time-independent) birth-death process with two possible
states (a binary character), with diversification rates
\(\{\lambda_0, \mu_0\}\) and \(\{\lambda_1, \mu_1\}\).</p>

<h4 id="clade-probabilities-d_n-i"><strong>Clade probabilities, $D_{N, i}$</strong></h4>

<p>We define $D_{N,0}(t)$ as the probability of observing lineage $N$
descending from a particular branch at time $t$, given that the lineage
at that time is in state 0. To compute the probability of observing the
lineage at some earlier time point, $D_{N,0}(t + \Delta t)$, we
enumerate all possible events that could occur within the interval
$\Delta t$. Assuming that $\Delta t$ is small—so that the probability of
more than one event occurring in the interval is negligible—there are
four possible scenarios within the time interval
(<a href="#bisse_events"></a>):</p>

<ol>
  <li>
    <p>nothing happens;</p>
  </li>
  <li>
    <p>a transition occurs, so the state changes $0 \rightarrow 1$;</p>
  </li>
  <li>
    <p>a speciation event occurs, and the right descendant subsequently
goes extinct before the present, or;</p>
  </li>
  <li>
    <p>a speciation event occurs and the left descendant subsequently goes
extinct before the present.</p>
  </li>
</ol>

<p>We are describing events within a branch of the tree (not at a node), so
for (3) and (4), we require that one of the descendant lineages go
extinct before the present because we do not observe a node in the tree
between $t$ and $t + \Delta t$.</p>

<figure id="bisse_events"><p><img src="figures/BiSSE_Events_D.png" /></p>
<figcaption>Possible events along a branch in the BiSSE model, used for deriving $D_{N,0}(t + \Delta t)$.
This is Figure 2 in <a class="citation" href="#Maddison2007">Maddison et al. (2007)</a>.</figcaption>
</figure>

<p>We can thus compute $D_{N,0}(t + \Delta t)$ as:</p>

\[\begin{aligned}
	D_{N,0}(t + \Delta t) = &amp; \;(1 - \mu_0 \Delta t) \times &amp; \text{in all cases, no extinction of the observed lineage} \\
&amp; \;[  (1 - q_{01} \Delta t)(1 - \lambda_0 \Delta t) D_{N,0}(t) &amp; \text{case (1) nothing happens} \\
&amp; \; + (q_{01} \Delta t) (1 - \lambda_0 \Delta t) D_{N,1}(t) &amp; \text{case (2) state change but no speciation} \\
&amp; \; + (1 - q_{01} \Delta t) (\lambda_0 \Delta t) E_0(t) D_{N,0}(t) &amp; \text{case (3) no state change, speciation, extinction} \\
&amp; \; + (1 - q_{01} \Delta t) (\lambda_0 \Delta t) E_0(t) D_{N,0}(t)] &amp; \text{case (4) no state change, speciation, extinction}
\end{aligned}\]

<p>A matching equation can be written down for $D_{N,1}(t+\Delta t)$.</p>

<p>To convert these difference equations into differential equations, we
take the limit $\Delta t \rightarrow 0$. With the notation that $i$ can
be either state 0 or state 1, and $j$ is the other state, this yields:</p>

\[\frac{\mathrm{d}D_{N,i}(t)}{\mathrm{d}t} = - \left(\lambda_i + \mu_i + q_{ij} \right) D_{N,i}(t) + q_{ij} D_{N,j}(t) + 2 \lambda_i E_i(t) D_{N,i}(t)
\tag{1}\label{eq:one}\]

<h3 id="extinction-probabilities-e_i">Extinction probabilities, $E_i$</h3>

<p>To solve the above equations for $D_{N, i}$, we see that we need the
extinction probabilities. Define $E_0(t)$ as the probability that a
lineage in state 0 at time $t$ goes extinct before the present. To
determine the extinction probability at an earlier point,
$E_0(t+\Delta t)$, we can again enumerate all the possible events in the
interval $\Delta t$ (<a href="#BiSSE_Events_E"></a>):</p>

<ol>
  <li>
    <p>the lineage goes extinct within the interval;</p>
  </li>
  <li>
    <p>the lineage neither goes extinct nor speciates, resulting in a
single lineage that must eventually go extinct before the present;</p>
  </li>
  <li>
    <p>the lineage neither goes extinct nor speciates, but there is a state
change, resulting in a single lineage that must go extinct before
the present, or;</p>
  </li>
  <li>
    <p>the lineage speciates in the interval, resulting in <em>two</em> lineages
that must eventually go extinct before the present.</p>
  </li>
</ol>

\[\begin{aligned}
E_0(t + \Delta t) = &amp;\; \mu_0\Delta t +
&amp; \text{case (1) extinction in the interval} \\
&amp; (1 - \mu_0\Delta t) \times &amp; \text{no extinction in the interval and \dots} \\
&amp; \;[(1-q_{01}\Delta t)(1-\lambda_0 \Delta t) E_0(t) &amp; \text{case (2) nothing happens, but subsequent extinction} \\
&amp; \;+ (q_{01}\Delta t) (1-\lambda_0 \Delta t) E_1(t) &amp; \text{case (3) state change and subsequent extinction} \\
&amp; \;+ (1 - q_{01} \Delta t) (\lambda_0 \Delta t) E_0(t)^2] &amp; \text{case (4) speciation and subsequent extinctions}
\end{aligned}\]

<p>Again, a matching equation for $E_1(t+\Delta t)$ can be written down.</p>

<figure id="BiSSE_Events_E"><p><img src="figures/BiSSE_Events_E.png" /></p>
<figcaption>Possible events along a branch in the BiSSE model, used for deriving $E_0(t + \Delta t)$.
This is Figure 3 in <a class="citation" href="#Maddison2007">Maddison et al. (2007)</a>.</figcaption>
</figure>

<p>To convert these difference equations into differential equations, we
again take the limit $\Delta t \rightarrow 0$:</p>

\[\frac{\mathrm{d}E_i(t)}{\mathrm{d}t} = \mu_i - \left(\lambda_i + \mu_i + q_{ij} \right)E_i(t) + q_{ij} E_j(t) + \lambda_i E_i(t)^2
\tag{2}\label{eq:two}\]

<h3 id="initial-values-tips-and-sampling">Initial values: tips and sampling</h3>

<p>The equations above describe how to get the answer at time
$t + \Delta t$ assuming we already have the answer at time $t$. How do
we start this process? The answer is with our character state
observations, which are generally the tip state values. If species $s$
has state $i$, then $D_{s,i}(0) = 1$ (probability is 1 at time 0 [the
present] because we observed it for sure) and $E_i(0) = 0$ (probability
0 of being extinct at the present). For all states other than $i$,
$D_{s,j}(0) = 0$ and $E_j(0) = 1$.</p>

<p>We can adjust these initial conditions to allow for incomplete sampling.
If a proportion $\rho$ of species are included on the tree, we would
instead set $D_{s,i}(0) = \rho$ (probability of having state $s$ <em>and</em>
of being on the tree) and $E_i(0) = 1-\rho$ (probability of absent, due
to sampling rather than extinction). This simple form of incomplete
sampling assumes that any species is equally likely to be on the tree
<a class="citation" href="#FitzJohn2009">(FitzJohn et al. 2009)</a>.</p>

<h3 id="at-nodes">At nodes</h3>

<p>Equations \eqref{eq:one} and \eqref{eq:two} are the BiSSE ODEs,
describing probabilities along the branches of a phylogeny. We also need
to specify what happens with the clade probabilities (the $D$s) at the
nodes of the tree. BiSSE assumes the ancestor (called $A$) and
descendants (called $N$ and $M$) have the same state
(<em>i.e.,</em> there is no cladogenetic character
change). The initial value for the ancestral branch going into a node
(at time $t_A$) is then the product of the final values for each of the
daughter branches coming out of that node, times the instantaneous
speciation rate (to account for the observed speciation event):</p>

\[D_{A, i}(t_A) = D_{N, i}(t_A) D_{M, i}(t_A) \lambda_i
\tag{3}\label{eq:three}\]

<h3 id="at-the-root">At the root</h3>

<p>After we integrate equations \eqref{eq:one} and \eqref{eq:two}
from the tips to the root, dealing with nodes along the way via
equation \eqref{eq:three}, we arrive at the root with the $D$ values
(called $D_{R, i}$), one for each state. These need to be combined
somehow to get the overall likelihood of the data:</p>

\[\text{Likelihood(tree, tip states | model)} = \sum_i D_{R, i} \, p_{R, i}\]

<p>What probability weighting, $p_{R, i}$ should be used for the possible
root states? Sometimes a fixed approach is used, assuming that the prior
root state probabilities are either all equal, or are the same as the
observed tip state frequencies, or are the equilibrium state frequencies
under the model parameters. These assumptions do not have a real basis,
however (unless there is some external data that supports them), and
they can cause trouble <a class="citation" href="#Goldberg2008">(Goldberg and Igić 2008)</a>. An alternative is to use the
BiSSE probabilities themselves to determine the root state weightings,
essentially adjusting the weightings to be most consistent with the data
and BiSSE parameters <a class="citation" href="#FitzJohn2009">(FitzJohn et al. 2009)</a>. Perhaps better is to treat the
weightings as unknown parameters to be estimated. These estimates are
usually quite uncertain, but in a Bayesian framework, one can treat the
$p_{R, i}$ as nuisance parameters and integrate over them.</p>

<h3 id="bisse-model-parameters-and-their-interpretation">BiSSE model parameters and their interpretation</h3>

<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Interpretation</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>$\Psi$</td>
      <td>Phylogenetic tree with divergence times</td>
    </tr>
    <tr>
      <td>$T$</td>
      <td>Root age</td>
    </tr>
    <tr>
      <td>$q_{01}$</td>
      <td>Rate of transitions from 0 to 1</td>
    </tr>
    <tr>
      <td>$q_{10}$</td>
      <td>Rate of transitions from 1 to 0</td>
    </tr>
    <tr>
      <td>$\lambda_0$</td>
      <td>Speciation rate for state 0</td>
    </tr>
    <tr>
      <td>$\mu_0$</td>
      <td>Extinction rate for state 0</td>
    </tr>
    <tr>
      <td>$\lambda_1$</td>
      <td>Speciation rate for state 1</td>
    </tr>
    <tr>
      <td>$\mu_1$</td>
      <td>Extinction rate for state 1</td>
    </tr>
  </tbody>
</table>

<h3 class="subsection" id="equations-for-the-multi-state-birth-death-process">Equations for the multi-state birth-death process</h3>
<hr class="subsection" />

<p>The entire derivation above can easily be expanded to accommodate an
arbitrary number of states <a class="citation" href="#FitzJohn2012">(FitzJohn 2012)</a>. The only extra piece is
summing over all the possible state transitions. The resulting
differential equations within the branches are:</p>

\[\begin{aligned}
\frac{\mathrm{d}D_{N,i}(t)}{\mathrm{d}t} &amp;= - \left(\lambda_i + \mu_i + \sum\limits_{j \neq i}^k q_{ij} \right)D_{N,i}(t) + \sum\limits_{j \neq i}^k q_{ij} D_{N,j}(t) + 2\lambda_iE_i(t)D_{N,i}(t) \\
\frac{\mathrm{d}E_i(t)}{\mathrm{d}t} &amp;= \mu_i - \left(\lambda_i + \mu_i + \sum\limits_{j \neq i}^k q_{ij} \right)E_i(t) + \sum\limits_{j \neq i}^k q_{ij} E_j(t) + \lambda_i E_i(t)^2
\end{aligned}\]

<ol class="bibliography"><li><span id="Beaulieu2016">Beaulieu J.M., O’Meara B.C. 2016. Detecting hidden diversification shifts in models of trait-dependent speciation and extinction. Systematic Biology. 65:583–601.</span>

<a href="https://doi.org/10.1093/sysbio/syw022">10.1093/sysbio/syw022</a>

</li>
<li><span id="FitzJohn2009">FitzJohn R.G., Maddison W.P., Otto S.P. 2009. Estimating trait-dependent speciation and extinction rates from incompletely resolved phylogenies. Systematic Biology. 58:595–611.</span>

<a href="https://doi.org/10.1093/sysbio/syp067">10.1093/sysbio/syp067</a>

</li>
<li><span id="FitzJohn2010">FitzJohn R.G. 2010. Quantitative Traits and Diversification. Systematic Biology. 59:619–633.</span>

<a href="https://doi.org/10.1093/sysbio/syq053">10.1093/sysbio/syq053</a>

</li>
<li><span id="FitzJohn2012">FitzJohn R.G. 2012. Diversitree: Comparative Phylogenetic Analyses of Diversification in R. Methods in Ecology and Evolution. 3:1084–1092.</span>

<a href="https://doi.org/10.1111/j.2041-210X.2012.00234.x">10.1111/j.2041-210X.2012.00234.x</a>

</li>
<li><span id="Freyman2018">Freyman W.A., Höhna S. 2018. Cladogenetic and anagenetic models of chromosome number evolution: a Bayesian model averaging approach. Systematic Biology. 67:1995–215.</span>

</li>
<li><span id="Goldberg2011">Goldberg E.E., Lancaster L.T., Ree R.H. 2011. Phylogenetic Inference of Reciprocal Effects between Geographic Range Evolution and Diversification. Systematic Biology. 60:451–465.</span>

<a href="https://doi.org/10.1093/sysbio/syr046">10.1093/sysbio/syr046</a>

</li>
<li><span id="Goldberg2008">Goldberg E.E., Igić B. 2008. On Phylogenetic Tests of Irreversible Evolution. Evolution. 62:2727–2741.</span>

<a href="https://doi.org/10.1111/j.1558-5646.2008.00505.x">10.1111/j.1558-5646.2008.00505.x</a>

</li>
<li><span id="Goldberg2012">Goldberg E.E., Igić B. 2012. Tempo and Mode in Plant Breeding System Evolution. Evolution. 66:3701–3709.</span>

<a href="https://doi.org/10.1111/j.1558-5646.2012.01730.x">10.1111/j.1558-5646.2012.01730.x</a>

</li>
<li><span id="Maddison2007">Maddison W.P., Midford P.E., Otto S.P. 2007. Estimating a binary character’s effect on speciation and extinction. Systematic Biology. 56:701.</span>

<a href="https://doi.org/10.1080/10635150701607033">10.1080/10635150701607033</a>

</li>
<li><span id="Maddison2006b">Maddison W.P. 2006. Confounding Asymmetries in Evolutionary Diversification and Character Change. Evolution. 60:1743–1746.</span>

<a href="https://doi.org/10.1111/j.0014-3820.2006.tb00517.x">10.1111/j.0014-3820.2006.tb00517.x</a>

</li>
<li><span id="MagnusonFord2012">Magnuson-Ford K., Otto S.P. 2012. Linking the Investigations of Character Evolution and Species Diversification. The American Naturalist. 180:225–245.</span>

<a href="https://doi.org/10.1086/666649">10.1086/666649</a>

</li>
<li><span id="Mitter1988">Mitter C., Farrell B., Wiegemann B. 1988. The Phylogenetic Study of Adaptive Zones: Has Phytophagy Promoted Insect Diversification? The American Naturalist. 132:107–128.</span>

<a href="https://doi.org/10.1086/284840">10.1086/284840</a>

</li>
<li><span id="Pagel1999">Pagel M. 1999. The Maximum Likelihood Approach to Reconstructing Ancestral Character States of Discrete Characters on Phylogenies. Systematic Biology. 48:612–622.</span>

<a href="https://doi.org/10.1080/106351599260184">10.1080/106351599260184</a>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
