<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Introduction to Diversification Rate Estimation</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Introduction to Diversification Rate Estimation</h1>
	<h3 class="subtitle">Overview of Analyses, Models and Theory</h3>
	<h4 class="authors">Sebastian Höhna</h4>
  <h5>Last modified on March 11, 2022</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/intro/">Getting Started with RevBayes and Rev Language Syntax</a></li>
          
            <li><a href="/tutorials/mcmc/">Introduction to Markov chain Monte Carlo (MCMC) Sampling</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





</div>
<h2 class="section" id="diversification_rate_overview">Overview: Diversification Rate Estimation</h2>
<hr class="section" />

<p>Stochastic branching models allow for inference of speciation and
extinction rates. In these tutorials we focus on the different types of macroevolutionary
models to study diversification processes and thus the
diversification-rate parameters themselves.</p>

<h3 class="subsection" id="types-of-hypotheses-for-estimating-diversification-rates">Types of Hypotheses for Estimating Diversification Rates</h3>
<hr class="subsection" />

<p>Macroevolutionary diversification rate estimation focuses on different key hypothesis,
which may include:
adaptive radiation, diversity-dependent and character diversification, key innovations, and
mass extinction.
We classify these hypotheses primarily into questions whether diversification rates vary through time,
and if so, whether some external, global factor has driven diversification rate changes, or if
diversification rates vary among lineages, and if so, whether some species specific factor is correlated with
the diversification rates.</p>

<p>Below, we describe each of the fundamental questions regarding diversification rates.</p>

<h4 class="subsubsection" id="1-constant-diversification-rate-estimation">(1) Constant diversification-rate estimation</h4>
<hr class="subsubsection" />

<p><em>What is the global rate of diversification in my phylogeny?</em>
The most basic models estimate parameters of the birth-death process
(<em>i.e.</em>, rates of speciation and extinction, or composite parameters
such as net-diversification and relative-extinction rates)
under the assumption that rates have remained constant across lineages and through time.
This is the most basic example and should be treated as a primer and introduction into the topic.</p>

<p>For more information, we recommend the <a href="/tutorials/divrate/simple.html">Simple Diversification Rate Estimation</a>.</p>

<h4 class="subsubsection" id="2-diversification-rate-variation-through-time">(2) Diversification rate variation through time</h4>
<hr class="subsubsection" />

<p><em>Is there diversification rate variation through time in my phylogeny?</em>
There are several reasons why diversification rates for the entire study group can vary through time, for example:
adaptive radiation, diversity dependence and mass-extinction events.
We can detect a signal any of these causes by detecting diversification rate variation through time.</p>

<p>The different tutorials references below cover different scenarios for diversification rate variation through time.
The common theme of these studies is that the diversification process is tree-wide, that is,
all lineages of the study group have the exact same rates at a given time.</p>

<h4 class="subsubsection" id="2a-detecting-diversification-rate-variation-through-time">(2a) Detecting diversification rate variation through time</h4>
<hr class="subsubsection" />

<p>In RevBayes we use an episodic birth-death model to study diversification rate variation through time.
That is, we assume that diversification rates are constant within an epoch but may shift between episodes (<a class="citation" href="#Stadler2011">Stadler (2011), Höhna (2015)</a>).
Then, we are estimating the diversification rates for each episode, and thus diversification rate variation through time.</p>

<p>You can find examples and more information in the <a href="/tutorials/divrate/ebd.html">Episodic Diversification Rate Estimation</a>.</p>

<h4 class="subsubsection" id="2b-detecting-the-impact-of-mass-extinction-events-on-diversification">(2b) Detecting the impact of mass-extinction events on diversification</h4>
<hr class="subsubsection" />

<p>Another question in this category asks whether our study tree was impacted by a mass-extinction event
(where a large fraction of the standing species diversity is suddenly lost, e.g., <a class="citation" href="#Hoehna2015a">Höhna (2015), May et al. (2016), Magee and Höhna (2021)</a>).
That is, we infer and test for the impact of instantaneous mass extinction events where each species alive at the given time has a probability of survival of the event.</p>

<p>You can find examples and more information in the <a href="/tutorials/divrate/efbdp_me.html">Mass Extinction Estimation</a>.</p>

<h4 class="subsubsection" id="2c-diversification-rate-correlation-to-environmental-eg-abiotic-factors">(2c) Diversification-rate correlation to environmental (e.g., abiotic) factors</h4>
<hr class="subsubsection" />

<p><em>Are diversification rates correlated with some abiotic (e.g., environmental) variable in my phylogeny?</em>
If we have found evidence in the previous section that diversification rates vary through time,
then we can start asking the question whether these changes in diversification
rates are driven by some abiotic (e.g., environmental) factors.
For example, we can ask whether changes in diversification rates are correlated with
environmental factors, such as environmental CO<sub>2</sub> or temperature <a class="citation" href="#Condamine2013">(Condamine et al. 2013; Condamine et al. 2018; Palazzesi et al. 2022)</a>.</p>

<p>You can find examples and more information in the <a href="/tutorials/divrate/env.html">Environmental-dependent Speciation &amp; Extinction Rates</a>.</p>

<h4 class="subsubsection" id="3-diversification-rate-variation-across-branches-estimation">(3) Diversification-rate variation across branches estimation</h4>
<hr class="subsubsection" />

<p><em>Is there diversification rate variation among lineages in my phylogeny?</em>
There are several reasons why diversification rates can vary among lineages primarily due to species specific factors (intrinsic and extrinsic), for example, key innovations.
First, we can try to detect a signal of rate variation among lineages, and then we can test if their are variables that are associated with this among lineage rate variation.
The different tutorials references below cover different scenarios for diversification rate variation among lineages.</p>

<h4 class="subsubsection" id="3a-detecting-diversification-rate-variation-across-branches-estimation">(3a) Detecting diversification-rate variation across branches estimation</h4>
<hr class="subsubsection" />

<p><em>Is there evidence that diversification rates have varied across the branches of my phylogeny?</em>
Have there been significant diversification-rate shifts along branches in my phylogeny, and if so, how many shifts, what magnitude of rate-shifts and along which branches?
Similarly, one may ask what are the branch-specific diversification rates?</p>

<p>You can study diversification rate variation among lineages using our birth-death-shift process <a class="citation" href="#Hoehna2019">(Höhna et al. 2019)</a>.
Examples and more information is provided in the <a href="/tutorials/divrate/branch_specific.html">Branch-Specific Diversification Rate Estimation</a>.</p>

<h4 class="subsubsection" id="3b-character-dependent-diversification-rate-estimation">(3b) Character-dependent diversification-rate estimation</h4>
<hr class="subsubsection" />

<p>If we have found that there is rate variation among lineage, then we could ask if diversification rates correlated with some biotic (e.g., morphological) variable.
This can be addressed by using character-dependent birth-death models
(often also called state-dependent speciation and extinction models; SSE models).
Character-dependent diversification-rate models aim to identify
overall correlations between diversification rates and organismal
features (binary and multi-state discrete morphological traits,
continuous morphological traits, geographic range, etc.). For example,
one can hypothesize that a binary character, say if an organism is
herbivorous/carnivorous or self-compatible/self-incompatible, impact the
diversification rates. Then, if the organism is in state 0 (e.g., is herbivorous) it has a lower (or higher) diversification rate than if the organism is in state 1 (e.g., carnivorous) <a class="citation" href="#Maddison2007">(Maddison et al. 2007)</a>.</p>

<p>You can find examples and more information in</p>
<ul>
  <li><a href="/tutorials/sse/bisse-intro.html">Background on state-dependent diversification rate estimation</a></li>
  <li><a href="/tutorials/sse/bisse.html">State-dependent diversification with BiSSE and MuSSE</a> tutorial</li>
  <li><a href="/tutorials/sse/hisse.html">State-dependent diversification with HiSSE</a> tutorial</li>
  <li><a href="/tutorials/sse/classe.html">State-dependent diversification with the ClaSSE model</a> tutorial</li>
  <li><a href="/tutorials/chromo/">Chromosome Evolution</a> tutorial</li>
</ul>

<h4 class="subsubsection" id="4-general-extension-to-diversification-rate-estimation">(4) General Extension to Diversification Rate Estimation</h4>
<hr class="subsubsection" />

<p>There exist some general considerations, assumptions and extensions that apply to most/all diversification rate models.
We provide a few general topics.</p>

<h4 class="subsubsection" id="4a-incomplete-taxon-sampling">(4a) Incomplete taxon sampling</h4>
<hr class="subsubsection" />

<p>For most study groups, we do not have all extant taxa sampled.
It is important that we properly model incomplete taxon sampling because otherwise our parameter estimates are biased <a class="citation" href="#Hoehna2011">(Höhna et al. 2011; Höhna 2014; Palazzesi et al. 2022)</a>.
You can find examples and more information in the <a href="/tutorials/divrate/sampling.html">Diversification Rate Estimation with Missing Taxa</a>.</p>

<h4 class="subsubsection" id="4b-conditions-of-the-birth-death-process">(4b) Conditions of the Birth-Death Process</h4>
<hr class="subsubsection" />

<p>As any statistical model, the birth-death process includes several assumptions/conditions.
Primarily, we condition the process if we only consider study groups that (a) survived until the present, (b) left exactly $N$ extant taxa, or (c) no restrictions.
The conditions become a bit more involved if phylogenies with fossils are considered.
You can find more discussion and examples in the <a href="/tutorials/divrate/conditions.html">Assumptions of Diversification Rate Estimation</a>.</p>

<h2 class="section" id="models">Diversification Rate Models</h2>
<hr class="section" />

<p>We begin this section with a general introduction to the stochastic
birth-death branching process that underlies inference of
diversification rates in RevBayes. This primer will
provide some details on the relevant theory of stochastic-branching
process models. We appreciate that some readers may want to skip this
somewhat technical primer; however, we believe that a better
understanding of the relevant theory provides a foundation for
performing better inferences. We then discuss a variety of specific
birth-death models, but emphasize that these examples represent only a
tiny fraction of the possible diversification-rate models that can be
specified in RevBayes.</p>

<h3 class="subsection" id="the-birth-death-branching-process">The birth-death branching process</h3>
<hr class="subsection" />

<figure id="fig_birth_death_shift"><p><img src="figures/BirthDeathShift.png" width="50%" height="50%" /></p>
<figcaption>A realization of the birth-death process with mass extinction.
Lineages that have no extant or sampled descendant are shown in gray and
surviving lineages are shown in a thicker black line.</figcaption>
</figure>

<p>Our approach is based on the <em>reconstructed evolutionary process</em>
described by <a class="citation" href="#Nee1994b">(Nee et al. 1994)</a>; a birth-death process in which only sampled,
extant lineages are observed. Let $N(t)$ denote the number of species at
time $t$. Assume the process starts at time $t_1$ (the ‘crown’ age of
the most recent common ancestor of the study group, $t_\text{MRCA}$)
when there are two species. Thus, the process is initiated with two
species, $N(t_1) = 2$. We condition the process on sampling at least one
descendant from each of these initial two lineages; otherwise $t_1$
would not correspond to the $t_\text{MRCA}$ of our study group. Each
lineage evolves independently of all other lineages, giving rise to
exactly one new lineage with rate $b(t)$ and losing one existing lineage
with rate $d(t)$ (<a href="#fig_birth_death_shift"></a> and
<a href="#fig_bdp"></a>). Note that although each lineage evolves
independently, all lineages share both a common (tree-wide) speciation
rate $b(t)$ and a common extinction rate $d(t)$
<a class="citation" href="#Nee1994b">(Nee et al. 1994; Höhna 2015)</a>. Additionally, at certain times,
$t_{\mathbb{M}}$, a mass-extinction event occurs and each species
existing at that time has the same probability, $\rho$, of survival.
Finally, all extinct lineages are pruned and only the reconstructed tree
remains (<a href="#fig_birth_death_shift"></a>).</p>

<figure id="fig_bdp"><p><img src="figures/birth-death-sketch.png" width="75%" height="75%" /></p>
<figcaption><strong>Examples of trees produced under a birth-death process.</strong>
The process is initiated at the first speciation event (the ‘crown-age’ of the MRCA)
when there are two initial lineages. At each speciation event the ancestral lineage is replaced by two
descendant lineages. At an extinction event one lineage simply
terminates. (A) A complete tree including extinct lineages. (B) The
reconstructed tree of tree from A with extinct lineages pruned away. (C)
A <em>uniform</em> subsample of the tree from B, where each species was sampled
with equal probability, $\rho$. (D) A <em>diversified</em> subsample of the
tree from B, where the species were selected so as to maximize diversity.</figcaption>
</figure>

<p>To condition the probability of observing the branching times on the
survival of both lineages that descend from the root, we divide by
$P(N(T) &gt; 0 | N(0) = 1)^2$. Then, the probability density of the
branching times, $\mathbb{T}$, becomes</p>

\[\begin{aligned}
P(\mathbb{T}) = \frac{\overbrace{P(N(T) = 1 \mid N(0) = 1)^2}^{\text{both initial lineages have one descendant}}}{ \underbrace{P(N(T) &gt; 0 \mid N(0) = 1)^2}_{\text{both initial lineages survive}} } \times \prod_{i=2}^{n-1} \overbrace{i \times b(t_i)}^{\text{speciation rate}} \times \overbrace{P(N(T) = 1 \mid N(t_i) = 1)}^\text{lineage has one descendant},
\end{aligned}\]

<p>and the probability density of the reconstructed tree (topology and branching times) is then</p>

\[\begin{aligned}
P(\Psi) = \; &amp; \frac{2^{n-1}}{n!(n-1)!} \times \left( \frac{P(N(T) = 1 \mid N(0) = 1)}{P(N(T) &gt; 0 \mid N(0) = 1)} \right)^2 \nonumber\\
		  \; &amp; \times \prod_{i=2}^{n-1} i \times b(t_i) \times P(N(T) = 1 \mid N(t_i) = 1)
		  \label{eq:tree_probability}
\end{aligned}\]

<p>We can expand Equation ([eq:tree_probability]) by substituting
$P(N(T) &gt; 0 \mid N(t) =1)^2 \exp(r(t,T))$ for
$P(N(T) = 1 \mid N(t) = 1)$, where $r(u,v) = \int^v_u d(t)-b(t)dt$; the
above equation becomes</p>

\[\begin{aligned}
P(\Psi) = \; &amp; \frac{2^{n-1}}{n!(n-1)!} \times \left( \frac{P(N(T) &gt; 0 \mid N(0) =1 )^2 \exp(r(0,T))}{P(N(T) &gt; 0 \mid N(0) = 1)} \right)^2 \nonumber\\
		  \; &amp; \times \prod_{i=2}^{n-1} i \times b(t_i) \times P(N(T) &gt; 0 \mid N(t_i) = 1)^2 \exp(r(t_i,T)) \nonumber\\
		= \; &amp; \frac{2^{n-1}}{n!} \times \Big(P(N(T) &gt; 0 \mid N(0) =1 ) \exp(r(0,T))\Big)^2 \nonumber\\
		  \; &amp; \times \prod_{i=2}^{n-1} b(t_i) \times P(N(T) &gt; 0 \mid N(t_i) = 1)^2 \exp(r(t_i,T)).
		\label{eq:tree_probability_substitution}
\end{aligned}\]

<p>For a detailed description of this substitution, see <a class="citation" href="#Hoehna2015a">Höhna (2015)</a>. Additional
information regarding the underlying birth-death process can be found in
<a class="citation" href="#Thompson1975">Thompson (1975)</a> [Equation 3.4.6] and <a class="citation" href="#Nee1994b">Nee et al. (1994)</a> for constant rates and
<a class="citation" href="#Hoehna2013">Höhna (2013), Höhna (2014), Höhna (2015)</a> for arbitrary rate functions.</p>

<p>To compute the equation above we need to know the rate function,
$r(t,s) = \int_t^s d(x)-b(x) dx$, and the probability of survival,
$P(N(T)!&gt;!0|N(t)!=!1)$.
<a class="citation" href="#Yule1925">Yule (1925)</a> and later <a class="citation" href="#Kendall1948">Kendall (1948)</a> derived the
probability that a process survives ($N(T) &gt; 0$) and the probability of
obtaining exactly $n$ species at time $T$ ($N(T) = n$) when the process
started at time $t$ with one species. Kendall’s results were summarized
in Equation (3) and Equation (24) in <a class="citation" href="#Nee1994b">Nee et al. (1994)</a></p>

\[\begin{aligned}
P(N(T)\!&gt;\!0|N(t)\!=\!1) &amp; = &amp; \left(1+\int\limits_t^{T} \bigg(\mu(s) \exp(r(t,s))\bigg) ds\right)^{-1} \label{eq:survival} \\ \nonumber \\
P(N(T)\!=\!n|N(t)\!=\!1) &amp; = &amp; (1-P(N(T)\!&gt;\!0|N(t)\!=\!1)\exp(r(t,T)))^{n-1} \nonumber\\
&amp; &amp; \times P(N(T)\!&gt;\!0|N(t)\!=\!1)^2 \exp(r(t,T)) \label{eq:N} %\\
%P(N(T)\!=\!1|N(t)\!=\!1) &amp; = &amp; P(N(T)\!&gt;\!0|N(t)\!=\!1)^2 \exp(r(t,T)) \label{eq:1}
\end{aligned}\]

<p>An overview for different diversification models is given in <a class="citation" href="#Hoehna2015a">Höhna (2015)</a>.</p>

<blockquote class="aside"><h2>Phylogenetic trees as observations</h2><p>The branching processes used here describe probability distributions on
phylogenetic trees. This probability distribution can be used to infer
diversification rates given an “observed” phylogenetic tree. In reality
we never observe a phylogenetic tree itself. Instead, phylogenetic trees
themselves are estimated from actual observations, such as DNA
sequences. These phylogenetic tree estimates, especially the divergence
times, can have considerable uncertainty associated with them. Thus, the
correct approach for estimating diversification rates is to include the
uncertainty in the phylogeny by, for example, jointly estimating the
phylogeny and diversification rates. For the simplicity of the following
tutorials, we take a shortcut and assume that we know the phylogeny
without error. For publication quality analysis you should always
estimate the diversification rates jointly with the phylogeny and
divergence times.</p>
</blockquote>

<ol class="bibliography"><li><span id="Condamine2013">Condamine F.L., Rolland J., Morlon H. 2013. Macroevolutionary perspectives to environmental change. Ecology Letters.</span>

<a href="https://doi.org/10.1111/ele.12062">10.1111/ele.12062</a>

</li>
<li><span id="Condamine2018">Condamine F.L., Rolland J., Höhna S., Sperling F.A.H., Sanmartín I. 2018. Testing the role of the Red Queen and Court Jester as drivers of the macroevolution of Apollo butterflies. Systematic Biology. 67:940–964.</span>

</li>
<li><span id="Hoehna2015a">Höhna S. 2015. The time-dependent reconstructed evolutionary process with a key-role for mass-extinction events. Journal of Theoretical Biology. 380:321–331.</span>

<a href="https://doi.org/http://dx.doi.org/10.1016/j.jtbi.2015.06.005">http://dx.doi.org/10.1016/j.jtbi.2015.06.005</a>

</li>
<li><span id="Hoehna2014a">Höhna S. 2014. Likelihood Inference of Non-Constant Diversification Rates with Incomplete Taxon Sampling. PLoS One. 9:e84184.</span>

<a href="https://doi.org/10.1371/journal.pone.0084184">10.1371/journal.pone.0084184</a>

</li>
<li><span id="Hoehna2013">Höhna S. 2013. Fast simulation of reconstructed phylogenies under global time-dependent birth-death processes. Bioinformatics. 29:1367–1374.</span>

<a href="https://doi.org/10.1093/bioinformatics/btt153">10.1093/bioinformatics/btt153</a>

</li>
<li><span id="Hoehna2019">Höhna S., Freyman W.A., Nolen Z., Huelsenbeck J.P., May M.R., Moore B.R. 2019. A Bayesian Approach for Estimating Branch-Specific Speciation and Extinction Rates. bioRxiv.</span>

<a href="https://doi.org/10.1101/555805">10.1101/555805</a>

</li>
<li><span id="Hoehna2011">Höhna S., Stadler T., Ronquist F., Britton T. 2011. Inferring speciation and extinction rates under different species sampling schemes. Molecular Biology and Evolution. 28:2577–2589.</span>

</li>
<li><span id="Kendall1948">Kendall D.G. 1948. On the Generalized "Birth-and-Death" Process. The Annals of Mathematical Statistics. 19:1–15.</span>

<a href="https://doi.org/10.1214/aoms/1177730285">10.1214/aoms/1177730285</a>

</li>
<li><span id="Maddison2007">Maddison W.P., Midford P.E., Otto S.P. 2007. Estimating a binary character’s effect on speciation and extinction. Systematic Biology. 56:701.</span>

<a href="https://doi.org/10.1080/10635150701607033">10.1080/10635150701607033</a>

</li>
<li><span id="Magee2021">Magee A.F., Höhna S. 2021. Impact of K-Pg Mass Extinction Event on Crocodylomorpha Inferred from Phylogeny of Extinct and Extant Taxa. bioRxiv.:426715.</span>

<a href="https://doi.org/10.1101/426715">10.1101/426715</a>

</li>
<li><span id="May2016">May M.R., Höhna S., Moore B.R. 2016. A Bayesian Approach for Detecting the Impact of Mass-Extinction Events on Molecular Phylogenies When Rates of Lineage Diversification May Vary. Methods in Ecology and Evolution. 7:947–959.</span>

<a href="https://doi.org/10.1111/2041-210X.12563">10.1111/2041-210X.12563</a>

</li>
<li><span id="Nee1994b">Nee S., May R.M., Harvey P.H. 1994. The Reconstructed Evolutionary Process. Philosophical Transactions: Biological Sciences. 344:305–311.</span>

<a href="https://doi.org/10.1098/rstb.1994.0068">10.1098/rstb.1994.0068</a>

</li>
<li><span id="Palazzesi2022">Palazzesi L., Hidalgo O., Barreda V.D., Forest F., Höhna S. 2022. The rise of grasslands is linked to atmospheric CO\textsubscript2 decline in the late Paleogene. Nature Communications. 13:293.</span>

</li>
<li><span id="Stadler2011">Stadler T. 2011. Mammalian phylogeny reveals recent diversification rate shifts. Proceedings of the National Academy of Sciences. 108:6187–6192.</span>

<a href="https://doi.org/10.1073/pnas.1016876108">10.1073/pnas.1016876108</a>

</li>
<li><span id="Thompson1975">Thompson E.A. 1975. Human evolutionary trees. Cambridge University Press Cambridge.</span>

</li>
<li><span id="Yule1925">Yule G.U. 1925. A mathematical theory of evolution, based on the conclusions of Dr. JC Willis, FRS. Philosophical Transactions of the Royal Society of London. Series B, Containing Papers of a Biological Character. 213:21–87.</span>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
