<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Tutorials</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <h1 class="maintitle">

RevBayes Tutorials

</h1>
<hr>

<p>This list shows all of the RevBayes tutorials for learning various aspects of RevBayes and Bayesian phylogenetic analysis.
Each one explicitly walks you through model specification and analysis set-up for different phylogenetic methods.
These tutorials have been written for new users to learn RevBayes at home, at workshops, and in courses taught at the undergraduate and graduate levels.
You may find that the styles are somewhat different between tutorials and that some  have overlapping content.</p>

<p>Please see the <a href="/tutorials/format">Tutorial Format</a> guide for details about how to read the tutorials.</p>

<p>Please see <a href="/tutorials/recommended">Recommended Software</a> for links to various software programs you may need to download in order to follow the tutorials.</p>

<p><a href="/developer/tutorial/" class="btn btn-warning" role="button">Contribute!</a></p>

<h3 id="intro">Introduction to RevBayes</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/intro/"><p>Getting Started with RevBayes and Rev Language Syntax</p>
</a>

<p class="subtitle"><p>Overview of introductory topics</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro/getting_started.html"><p>Getting Started with RevBayes</p>
</a>

<p class="subtitle"><p>A very basic overview on how to use RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro/rev.html"><p>Rev Language Syntax</p>
</a>

<p class="subtitle"><p>A very short introduction to the Rev language</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro/graph_models.html"><p>Introduction to Graphical Models</p>
</a>

<p class="subtitle"><p>A gentle introduction to graphical models, probabilistic programming, and MCMC using a simple linear regression example.</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/tutorial_structure/"><p>Setting Up RevBayes</p>
</a>

<p class="subtitle"><p>Preparing your computer to work with phylogenetic data in RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro/revgadgets.html"><p>Introduction to RevGadgets</p>
</a>

<p class="subtitle"><p>Plotting the output of RevBayes analyses in the R package RevGadgets.</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/docker/"><p>Using RevBayes with Docker</p>
</a>

<p class="subtitle"><p>Using a platform-independent container to run RevBayes on your computer or a computing cluster</p>
</p>
</div>



</div>

<h3 id="mcmc">Introduction to MCMC</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/mcmc/"><p>Introduction to Markov chain Monte Carlo (MCMC) Sampling</p>
</a>

<p class="subtitle"><p>Overview of topics on MCMC methods in RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc/poisson.html"><p>Introduction to MCMC</p>
</a>

<p class="subtitle"><p>A simple Poisson regression example</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc/binomial.html"><p>Introduction to MCMC using RevBayes</p>
</a>

<p class="subtitle"><p>Introduction to MCMC Simulation using a simple Binomial Model</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc/archery.html"><p>Introduction to MCMC using RevBayes</p>
</a>

<p class="subtitle"><p>A simple Archery example for building a hierarchical model and sampling under Markov Chain Monte Carlo</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc/moves.html"><p>Overview of Moves in MCMC</p>
</a>

<p class="subtitle"><p>Guidelines and Recommendation</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/convergence/"><p>Convergence assessment</p>
</a>

<p class="subtitle"><p>Phylogenetic convergence assessment using the R package Convenience</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc_troubleshooting/"><p>Debugging your Markov chain Monte Carlo (MCMC)</p>
</a>

<p class="subtitle"><p>Practical guidelines for issues with MCMC for Bayesian phylogenetic inference</p>
</p>
</div>



</div>

<h3 id="model_test">Model Selection and Testing</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_intro.html"><p>General Introduction to Model selection</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_subst_model.html"><p>Model selection of common substitution models for one locus</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_partition_model.html"><p>Model selection of partition models</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_subst_model_rj.html"><p>Model averaging of substitution models</p>
</a>

<p class="subtitle"><p>Reversible-jump MCMC over substitution models</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/intro_posterior_prediction/"><p>Introduction to Posterior Prediction</p>
</a>

<p class="subtitle"><p>Assessing the fit of Normal distributions to trait data</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_testing_pps/pps_data.html"><p>Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</p>
</a>

<p class="subtitle"><p>Model adequacy testing using posterior prediction (Data Version).</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_testing_pps/pps_inference.html"><p>Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</p>
</a>

<p class="subtitle"><p>Model adequacy testing using posterior prediction (Inference Version).</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/pps_morpho/"><p>Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</p>
</a>

<p class="subtitle"><p>Model adequacy testing using posterior prediction (Morphological Models).</p>
</p>
</div>












</div>

<h3 id="phylo_basic">Standard tree inference</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/dice/"><p>Understanding Continuous-Time Markov Models</p>
</a>

<p class="subtitle"><p>Simulating DNA sequence evolution with a die</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/ctmc/"><p>Nucleotide substitution models</p>
</a>

<p class="subtitle"><p>Phylogenetic inference of nucleotide data using RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/partition/"><p>Partitioned data analysis</p>
</a>

<p class="subtitle"><p>Current Protocols in Bioinformatics - Phylogenetic Inference using RevBayes (Protocol #2)</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_tree/V2.html"><p>Discrete morphology - Multistate Characters</p>
</a>

<p class="subtitle"><p>Phylogenetic inference with discrete morphological data</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_tree/"><p>Discrete morphology - Tree Inference</p>
</a>

<p class="subtitle"><p>Phylogenetic inference with discrete morphological data</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/dating/"><p>Dating trees</p>
</a>

<p class="subtitle"><p>Estimating species divergence times using RevBayes</p>
</p>
</div>






















<div class="tutorial ">

<a class="title" href="/tutorials/clocks/"><p>Relaxed Clocks &amp; Time Trees</p>
</a>

<p class="subtitle"><p>Comparing relaxed clock models &amp; estimating rooted time trees</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sequential_bayes/"><p>Sequential Bayesian inference of phylogeny</p>
</a>

<p class="subtitle"><p>Example sequential Bayesian analysis for divergence time estimation in RevBayes</p>
</p>
</div>
















<div class="tutorial ">

<a class="title" href="/tutorials/pomobalance/"><p>Polymorphism-aware phylogenetic models with balancing selection</p>
</a>

<p class="subtitle"><p>Species tree inference and identification of preferred allele frequency for balancing selection in RevBayes</p>
</p>
</div>
















<div class="tutorial ">

<a class="title" href="/tutorials/relative_time_constraints/"><p>Dating with Relative Constraints</p>
</a>

<p class="subtitle"><p>Constraining the relative order of nodes when dating a phylogeny</p>
</p>
</div>



</div>

<h3 id="phylo_complex">Complex hierarchical models for phylogenetic inference</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/fbd/fbd_specimen.html"><p>Combined-Evidence Analysis and the Fossilized Birth-Death Process for Analysis of Extant Taxa and Fossil Specimens</p>
</a>

<p class="subtitle"><p>Joint inference of divergence times and phylogenetic relationships of fossil and extant taxa</p>
</p>
</div>






</div>

<h3 id="phylo_div_rate">Discrete Morphology</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/"><p>Introduction to Discrete Morphology Evolution</p>
</a>

<p class="subtitle"><p>Overview of types of Analyses</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/ase.html"><p>Discrete morphology - Ancestral State Estimation</p>
</a>

<p class="subtitle"><p>The Equal Rates Markov Model</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/ase_free.html"><p>Discrete morphology - Ancestral State Estimation with the independent rates model</p>
</a>

<p class="subtitle"><p>Testing for Unequal Rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/ase_irreversible.html"><p>Discrete morphology - Ancestral State Estimation and Irreversibility</p>
</a>

<p class="subtitle"><p>Testing for Irreversibility</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/scm_hrm.html"><p>Discrete morphology - Stochastic Character Mapping and Hidden Rates</p>
</a>

<p class="subtitle"><p>Mapping Character Transitions and Testing for Rate Variation</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/corr.html"><p>Discrete morphology - Correlation among Characters</p>
</a>

<p class="subtitle"><p>Testing for Correlation</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph_ase/ase_mammals.html"><p>Discrete morphology - Ancestral State Estimation (Mammals &amp; Placenta Type)</p>
</a>

<p class="subtitle"><p>Ancestral State Estimation and Testing for Irreversibility</p>
</p>
</div>



</div>

<h3 id="phylo_trait">Comparative methods</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/cont_trait_intro.html"><p>Introduction to Models of Continuous-Character Evolution</p>
</a>

<p class="subtitle"><p>Overview of Standard Models of Continuous-Character Evolution</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/simple_bm.html"><p>Simple Brownian Rate Estimation</p>
</a>

<p class="subtitle"><p>Estimating rates of Brownian-motion evolution</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/relaxed_bm.html"><p>Relaxed Brownian Rate Estimation</p>
</a>

<p class="subtitle"><p>Estimating branch-specific rates of Brownian-motion evolution</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/state_dependent_bm.html"><p>State-Dependent Brownian Rate Estimation</p>
</a>

<p class="subtitle"><p>Estimating state-dependent rates of Brownian-motion evolution</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/multivariate_bm.html"><p>Multivariate Brownian Motion</p>
</a>

<p class="subtitle"><p>Accounting for correlations among continuous traits</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/simple_ou.html"><p>Simple Ornstein-Uhlenbeck Models</p>
</a>

<p class="subtitle"><p>Estimating optima under Ornstein-Uhlenbeck evolution</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/cont_traits/relaxed_ou.html"><p>Relaxed Ornstein-Uhlenbeck Models</p>
</a>

<p class="subtitle"><p>Estimating lineage-specific optima under Ornstein-Uhlenbeck evolution</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/host_rep/host_rep.html"><p>Reconstructing Host Repertoire Evolution</p>
</a>

<p class="subtitle"><p>Inference of ancestral ecological interactions</p>
</p>
</div>






</div>

<h3 id="phylo_biogeo">Diversification Rate Estimation</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/divrate/div_rate_intro.html"><p>Introduction to Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Overview of Analyses, Models and Theory</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/simple.html"><p>Simple Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Comparing different constant-rate models of lineage diversification</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/ebd.html"><p>Episodic Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Estimating Rate-Variation Through Time in Diversification Rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/env.html"><p>Environmental-dependent Speciation &amp; Extinction Rates</p>
</a>

<p class="subtitle"><p>Estimating Correlation between Diversification Rates and Environmental Characters</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/efbdp_me.html"><p>Mass Extinction Estimation</p>
</a>

<p class="subtitle"><p>Estimating Mass Extinctions from Phylogenies with Fossil and Extant Taxa</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/branch_specific.html"><p>Branch-Specific Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>How to estimate branch-specific shifts in diversification rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/bisse-intro.html"><p>Background on state-dependent diversification rate estimation</p>
</a>

<p class="subtitle"><p>An introduction to inference using state-dependent speciation and extinction (SSE) branching processes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/bisse.html"><p>State-dependent diversification with BiSSE and MuSSE</p>
</a>

<p class="subtitle"><p>Inference using the binary/multiple state-dependent speciation and extinction (BiSSE/MuSSE) branching process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/hisse.html"><p>State-dependent diversification with HiSSE</p>
</a>

<p class="subtitle"><p>Inference using the hidden character binary/multiple state-dependent speciation and extinction with (HiSSE) branching process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/classe.html"><p>State-dependent diversification with the ClaSSE model</p>
</a>

<p class="subtitle"><p>Accommodating both cladogenetic and anagenetic changes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/conditions.html"><p>Assumptions of Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Choices and impact of conditions of the birth-death process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/sampling.html"><p>Diversification Rate Estimation with Missing Taxa</p>
</a>

<p class="subtitle"><p>How to estimate diversification rates with incomplete taxon sampling</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/chromo/"><p>Chromosome Evolution</p>
</a>

<p class="subtitle"><p>Modeling chromosome evolution with ChromEvol, BiChroM, and ChromoSSE</p>
</p>
</div>



</div>

<h3 id="demographic">Biogeography</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_intro.html"><p>Introduction to phylogenetic biogeography with the DEC model</p>
</a>

<p class="subtitle"><p>Overview of the Dispersal-Extinction-Cladogenesis (DEC) model</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_simple.html"><p>Simple phylogenetic analysis using the DEC model</p>
</a>

<p class="subtitle"><p>Estimating ancestral ranges using the Dispersal-Extirpation-Cladogenesis (DEC) model</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_epoch.html"><p>Advanced phylogenetic analysis using the DEC model</p>
</a>

<p class="subtitle"><p>Ancestral range estimates using DEC with time- and region-dependent effects on dispersal rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_dating.html"><p>Biogeographic dating using the DEC model</p>
</a>

<p class="subtitle"><p>Estimating divergence times with molecular, biogeographic, and paleogeographic evidence with the DEC model</p>
</p>
</div>
















<div class="tutorial ">

<a class="title" href="/tutorials/fig_intro/"><p>Introduction to phylogenetic biogeography with the FIG model</p>
</a>

<p class="subtitle"><p>Using the feature-informed geographic state-dependent speciation-extinction (FIG) model for phylogenetic biogeography</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/multifig/"><p>Feature-informed geographic state-dependent speciation-extinction (FIG) model</p>
</a>

<p class="subtitle"><p>A phylogenetic model where GeoSSE event rates are informed by regional features</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/geosse/"><p>Geographic state-dependent speciation-extinction (GeoSSE) model</p>
</a>

<p class="subtitle"><p>A phylogenetic model of biogeographic dispersal, extinction, and speciation</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/timefig_simple/"><p>Time-heterogeneous FIG (TimeFIG) model</p>
</a>

<p class="subtitle"><p>A FIG model with paleogeographically varying regional features</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/timefig_dating/"><p>Biogeographic dating using the TimeFIG model</p>
</a>

<p class="subtitle"><p>Using the TimeFIG model to jointly infer divergence times and biogeographic history</p>
</p>
</div>



</div>

<h3>Demographic Inference</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/coalescent/"><p>Coalescent Analyses</p>
</a>

<p class="subtitle"><p>Estimating Demographic Histories with Bayesian Coalescent Skyline Plot Models</p>
</p>
</div>





































<div class="tutorial ">

<a class="title" href="/tutorials/stairwayplot/"><p>StairwayPlot Analyses</p>
</a>

<p class="subtitle"><p>Estimating Demographic Histories from SNP data with Bayesian StairwayPlot Approach</p>
</p>
</div>



</div>

<h3>Miscellaneous Tutorials</h3>

<div class="tutorialbox">




</div>


      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "default highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "default highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
