<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Tutorials</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
        <li><class="header-search">
  <form class="header-search-form" action="/search.html" method="get">
    <input type="text" id="search-box-nav" placeholder="Search..." name="query" style="border-color:black;background-color:white;height:32px;">
    <button type="submit" style="color:black; border-color:black;width:32px;height:32px;padding-top:4px">
        <img src="https://revbayes.github.io/assets/img/search.png" height="22px" width="28px"/>
    </button>
  </form>
</div>

</li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <!-- adds margin to main content -->
      <div class="container" style="padding-left:20px padding-right:20px;">
          <h1 class="maintitle">

Tutorial modules

</h1>
<hr>

<p>This page is intended to guide users in assembling a set of tutorials that will help them learn 
how to use RevBayes for a given analysis. Each module lists the tutorials that focus on a particular 
topic, as well as those that will provide prerequisite knowledge or resources for deeper learning.  <br />
We encourage users with no RevBayes experience to start with the <strong>RevBayes fundamentals</strong> module.</p>

<p>To see all available tutorials, including the ones for which modules are not currently written, 
go to the <a href="/all_tutorials">full tutorials page</a>.</p>

<blockquote class="aside"><h2>RevBayes fundamentals</h2><p>This module is for users who have no experience with RevBayes. The tutorials contained here walk
the user through the basics of installing and using RevBayes, and some of the theory behind
RevBayes analyses.</p>

<ul>
  <li>Pre-requisites: Basic knowledge on Bayesian statistics.</li>
</ul>

<ol>
  <li><a href="/tutorials/intro/getting_started">Getting started with RevBayes</a>. This tutorial includes getting RevBayes working on your machine, and learning about the fundamentals of setting up models in the software.</li>
  <li>MCMC fundamentals. Three tutorials lay out the basics of Markov Chain Monte Carlo analysis in RevBayes, using three different examples: <a href="/tutorials/mcmc/poisson">Airline fatalities</a>, <a href="/tutorials/mcmc/binomial">coin flips</a>, and <a href="/tutorials/mcmc/archery">archery</a>. It is recommended that absolute beginners go through all three.</li>
</ol>

<p>By the end of this module, you should be able to build RevBayes, write simple Rev code, and build 
simple models on which to run MCMC analyses. From here, you can select the module(s) below that
best represent your interests and intended analyses.</p>
</blockquote>

<blockquote class="aside"><h2>MCMC assessment</h2><p>This module is for users looking to debug or analyze their MCMC results deeper. The tutorials will walk the user through best practices when analyzing posterior samples, and tips on how to improve your analyses if your analysis has not converged.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>, and results from some analysis (e.g. from completion of one of the modules below).</li>
</ul>

<ol>
  <li>Assessing convergence of your MCMC results: <a href="/tutorials/convergence">Convergence assessment</a>. This tutorial will walk you through theory and examples on how to assess the convergence of your MCMC chain.</li>
  <li>Practical exercises on assessing convergence: <a href="/tutorials/mcmc_troubleshooting">Debugging your MCMC</a>. This tutorial contains more exercises on assessing convergence, and provides tips on how to improve convergence in your analysis.</li>
</ol>

<ul>
  <li><a href="/tutorials/intro/revgadgets">Introduction to RevGadgets</a> walks you through the use of RevGadgets, an R package that is an optional but convenient tool for postprocessing of RevBayes results.</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Molecular phylogenetic inference</h2><p>This module is intended for users looking to infer a phylogenetic tree using molecular (DNA) data.
Users will learn how to set up basic molecular evolution model, and be presented with more advanced
tutorials to tailor their learning to their desired analyses.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/dice">Understanding continuous-time Markov models</a>. This tutorial will get you up to speed with the theory behind molecular evolution models.</li>
  <li>Basic molecular phylogenetic inference: <a href="/tutorials/ctmc">Nucleotide substitution models</a>. This tutorial will walk you through a simple example of molecular phylogeny inference.</li>
</ol>

<ul>
  <li>If interested in partitioning molecular data: <a href="/tutorials/partition">Partitioned data analysis</a>.</li>
  <li>If interested in choosing between models of molecular evolution:
    <ol>
      <li><a href="/tutorials/model_selection_bayes_factors/bf_intro">General introduction to model selection</a>.</li>
      <li><a href="/tutorials/model_selection_bayes_factors/bf_subst_model">Model selection for one locus</a>.</li>
      <li><a href="/tutorials/model_selection_bayes_factors/bf_partition_model">Model selection for partition models</a>.</li>
    </ol>
  </li>
</ul>

<p>By the end of this module, you should be able to infer a phylogenetic tree with a simple molecular 
evolution model. Depending on your interests, you will also have learned about how to set up
partitioned models, and choose the best models for your dataset.</p>
</blockquote>

<blockquote class="aside"><h2>Polymorphism-aware molecular phylogenetics</h2><p>This module is intended for users looking to infer a phylogenetic tree using molecular data including polymorphisms. Users will learn how to modify their molecular evolution models to account for the presence of polymorphisms in the dataset, including, if desired, balanced selection. Note that the tutorials in this module use code on a separate RevBayes branch, requiring users to run <code class="language-plaintext highlighter-rouge">git checkout dev_PoMo_SNP</code> within their local RevBayes repo before running analyses.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>, <strong>Molecular phylogenetic inference</strong>.</li>
</ul>

<ol>
  <li><a href="/tutorials/pomos">Polymorphism-aware phylogenetic models</a>.
    <ul>
      <li>If interested in including balancing selection in your model: <a href="/tutorials/pomobalance">Polymorphism-aware phylogenetic models with balancing selection</a>.</li>
    </ul>
  </li>
</ol>
</blockquote>

<blockquote class="aside"><h2>Morphological phylogenetic inference</h2><p>This module is intended for users looking to infer a phylogenetic tree using morphological data.
Users will learn how to set up basic morphological evolution models, including for characters
with more than two states.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/dice">Understanding continuous-time Markov models</a>. This tutorial will get you up to speed with the theory behind morphological evolution models.</li>
  <li>Morphological phylogenetics with binary traits: <a href="/tutorials/morph_tree">Tree inference with discrete morphology</a>. This tutorial will walk you through a simple example of morphological phylogeny inference, using only binary (i.e. two-state) characters.</li>
  <li>Expanding your dataset: <a href="/tutorials/morph_tree/V2">Multistate characters</a>. This tutorial will teach you how to accommodate characters with more than two states in your morphological phylogenetics analyses.</li>
</ol>
</blockquote>

<blockquote class="aside"><h2>Dating a phylogeny</h2><p>This module is intended for users looking to date a time-calibrated phylogenetic tree. Users will learn how to set up analyses using both molecular clocks and fossil data for node dating.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>, and either <strong>Molecular phylogenetic inference</strong> or <strong>Morphological phylogenetic inference</strong>. Note that while the tutorials below use molecular models, it is straightforward to adapt the clock model to a morphological analysis.</li>
</ul>

<ol>
  <li>Master dating tutorial: <a href="/tutorials/dating">Dating trees</a>. This tutorial includes:
    <ul>
      <li><a href="/tutorials/dating/global">Global molecular clocks</a>: The basics of molecular clocks.</li>
      <li><a href="/tutorials/dating/relaxed">Uncorrelated relaxed clocks</a>: More complex (and biologically realistic) molecular clocks.</li>
      <li><a href="/tutorials/dating/nodedate">Node dating with fossils</a>: Including fossil data for node dating. The master dating tutorial also links to a more advanced tip-dating tutorial (see the <strong>Total-evidence analysis with tip-dating</strong> module).</li>
    </ul>
  </li>
  <li>Bringing it all together: <a href="/tutorials/clocks">Relaxed clocks &amp; time trees</a>. This tutorial walks the user through a full analysis, including performing model selection between multiple clock models.</li>
</ol>

<ul>
  <li>If interested in enforcing the relative order of nodes in your analysis: <a href="/tutorials/relative_time_constraints">Dating with relative constraints</a>.</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Estimating diversification rates</h2><p>This module is intended for users looking to infer rates of speciation and extinction. Users will learn how to set up analyses using models in the birth-death model family, including the use of fossil data. While most of the tutorials here assume a fixed phylogenetic tree, some recommendations are made to point the user towards the direction of joint estimation as well.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/divrate/div_rate_intro">Introduction to diversification rate estimation</a>. This tutorial will walk you through the theory behind the birth-death models used in the following tutorials.</li>
  <li>Constant-rate models: <a href="/tutorials/divrate/simple">Simple diversification-rate estimation</a>. This tutorial will walk you through the basics of setting up a birth-death model for estimating rates of speciation and extinction from a fixed phylogenetic tree.</li>
</ol>

<p>The following tutorials are to be picked based on your intended analyses, as they lay out different hypotheses regarding diversification dynamics and how to set up a BD model to test those.</p>
<ul>
  <li>If interested in time-heterogeneous (or environmentally-dependent) diversification dynamics: <a href="/tutorials/divrate/ebd">Episodic diversification rate estimation</a>, <a href="/tutorials/divrate/env">Environmental-dependent speciation and extinction rates</a>.</li>
  <li>If interested in mass-extinction estimation, including fossil taxa: <a href="/tutorials/divrate/efbdp_me">Mass-extinction estimation</a>.</li>
  <li>If interested in jointly estimating topology and diversification rates: <a href="/tutorials/clocks">Relaxed clocks &amp; time-trees</a> (consider completing the <strong>Molecular phylogenetic inference</strong> and <strong>Dating a phylogeny</strong> modules).</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Total-evidence analysis with tip-dating</h2><p>This module is intended for users looking to infer a time-calibrated phylogenetic tree with both molecular and fossil data, and fossil ages.</p>

<ul>
  <li>Pre-requisites: <strong>Dating a phylogeny</strong> (and both <strong>Molecular phylogenetic inference</strong> and <strong>Morphological phylogenetic inference</strong>), and <strong>Estimating diversification rates</strong>.</li>
</ul>

<ol>
  <li><a href="/tutorials/fbd/fbd_specimen">The fossilized birth-death model</a>. This tutorial will push you to put together the skills learned in many previous tutorials, including setting up molecular and morphological evolution, clock, and birth-death models.</li>
</ol>

<ul>
  <li>If interested in more complex diversification dynamics, explore the advanced tutorials in <strong>Estimating diversification rates</strong>. Note that <code class="language-plaintext highlighter-rouge">dnFBDP</code> should be used instead of <code class="language-plaintext highlighter-rouge">dnBDP</code>/<code class="language-plaintext highlighter-rouge">dnEBDP</code>.</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>State-dependent diversification-rate estimation</h2><p>This module is intended for users looking to infer trait-dependent diversification rates using the state-dependent speciation and extinction (SSE) model family. As with <strong>Estimating diversification rates</strong>, the tutorials assume a fixed tree, but tips are included for joint estimation.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>, and <strong>Estimating diversification rates</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/sse/bisse-intro">Background on state-dependent diversification-rate estimation</a>. This tutorial will introduce you to the theory behind the SSE model family, building from the fundamentals on BD models you learned on the <strong>Estimating diversification rates</strong> module.</li>
  <li>Testing the effects of discrete traits on diversification: <a href="/tutorials/sse/bisse">State-dependent diversification with BiSSE and MuSSE</a>. This tutorial will walk you through the setup of the simplest SSE models: binary and multiple SSE (BiSSE and MuSSE), which are used to test the effects of binary and multiple state discrete traits on diversification, respectively.</li>
</ol>

<ul>
  <li>If interested in testing whether unobserved effects might impact diversification for your data: <a href="/tutorials/sse/hisse">State-dependent diversification with hidden SSE (HiSSE)</a>.</li>
  <li>If interested in the presence of cladogenetic state transitions: <a href="/tutorials/sse/classe">State-dependent diversification with cladogenetic SSE (ClaSSE)</a>.</li>
  <li>If interested in the effect of chromosome number on diversification: <a href="/tutorials/chromo">Chromosome evolution</a>.</li>
  <li>If interested in jointly estimating topology and state-dependent diversification rates, adapt <a href="/tutorials/clocks">Relaxed clocks &amp; time-trees</a> to use <code class="language-plaintext highlighter-rouge">dnCDBDP</code> or <code class="language-plaintext highlighter-rouge">dnGLHBDSP</code> instead of <code class="language-plaintext highlighter-rouge">dnBDP</code> (consider completing the <strong>Molecular phylogenetic inference</strong> and <strong>Dating a phylogeny</strong> modules).</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Ancestral-state estimation</h2><p>This module is intended for users looking to estimate the ancestral state of species in a fixed phylogenetic tree. Tutorials assume a fixed tree, and no diversification-rate estimation, but tips in that direction are provided.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/morph_ase">Introduction to Discrete Morphology Evolution</a>. This tutorial will introduce you to the ancestral-state estimation (ASE) exercises in RevBayes.</li>
  <li><a href="/tutorials/morph_ase/ase">Simple ASE</a>. This tutorial will walk you through the set up and analysis of a simple ASE model, with equal transition rates between all states.</li>
  <li>Independent rates: <a href="/tutorials/morph_ase/ase_free">ASE with the independent-rates model</a>. This tutorial builds on the previous one by including the possibility that rates are independent, providing more biological realism to your model. You can easily modify this model to allow only a subset of the rates to be free to vary.</li>
  <li>Irreversibility: <a href="/tutorials/morph_ase/ase_irreversible">ASE and irreversibility</a>. This tutorial adds one more piece of complexity that is often useful in ASE modelsâ€“irreversibility, <em>i.e.</em> ensuring some state transitions are not reversible.</li>
  <li>Worked example: <a href="/tutorials/morph_ase/ase_mammals">Mammals and placenta type</a>. This tutorial provides an example of ASE with irreversibility for the estimation of placenta types in mammals. As an exercise, try modifying the script to allow to make the free-rates model into an independent-rates model.</li>
</ol>

<ul>
  <li>If interested in testing for hidden effects in your model and/or estimating the timing of transitions: <a href="/tutorials/morph_ase/scm_hrm">Stochastic character mapping and hidden rates</a>.</li>
  <li>If interested in testing for correlations between characters: <a href="/tutorials/morph_ase/corr">Correlations among characters</a>.</li>
  <li>If interested in jointly estimating diversification rates and ancestral states, see the <strong>State-dependent diversification-rate estimation</strong> module. If interested in jointly estimating the topology, see the last bullet point in that module.</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Continuous trait evolution</h2><p>This module is intended for users looking to estimate rates of evolution for continuous traits. The tutorials assume a fixed phylogenetic tree. While ways to jointly estimate diversification-rates and/or topology with evolutionary rates are possible, there is no current tutorial that applies that in a straightforward manner. If you are interested in setting up such an analysis, please start a discussion at our <a href="https://github.com/revbayes/revbayes/discussions">GitHub Discussions</a> page for guidance.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>.</li>
</ul>

<ol>
  <li>Theoretical background: <a href="/tutorials/cont_traits/cont_trait_intro">Introduction to Models of Continuous-Character Evolution</a>. This tutorial will introduce you to the continuous-trait evolutionary-rate estimation exercises in RevBayes.</li>
  <li><a href="/tutorials/cont_traits/simple_bm">Simple Brownian rate estimation</a>. This tutorial will walk you through a simple analysis of continuous-trait rate estimation assuming Brownian motion, <em>i.e.</em> assuming the trait is evolving randomly, with no directional trend.</li>
  <li><a href="/tutorials/cont_traits/simple_ou">Ornstein-Uhlenbeck models</a>. This tutorial will walk you through a slightly more complicated example, using an Ornstein-Uhlenbeck (OU) model instead of a Brownian motion model. OU models expand Brownian motion models in assuming an optimal trait value, such that evolution towards that optimum is more likely than evolution away from it.</li>
</ol>

<ul>
  <li>If interested in relaxed clock models, with different branches of the phylogeny having different evolutionary rates: <a href="/tutorials/cont_traits/relaxed_bm">Relaxed Brownian rate estimation</a> and <a href="/tutorials/cont_traits/relaxed_ou">Relaxed OU models</a>.</li>
  <li>If interested in multiple continuous traits: <a href="/tutorials/cont_traits/multivariate_bm">Multivariate Brownian motion</a>.</li>
  <li>If interested in making Brownian rates state-dependent: <a href="/tutorials/cont_traits/state_dependent_bm">State-dependent Brownian rate estimation</a>.</li>
</ul>
</blockquote>

<blockquote class="aside"><h2>Biogeography</h2><p>This module is for researchers interested in modeling how species disperse, go extinct
and speciate over geological time. It includes examples for how to use biogeographic
models to reconstruct ancestral species ranges, test relationships between regional
features and biogeographic rates, and use paleogeographic information to time-calibrate
phylogenetic divergence times. Biogeographic models explored are Dispersal-Extinction-Cladogenesis (DEC),
Geographic State-dependent Speciation Extinction (GeoSSE), and Feature-Informed GeoSSE (FIG) models.</p>

<ul>
  <li>Pre-requisites: <strong>RevBayes fundamentals</strong>, <strong>Molecular phylogenetic analysis</strong>, and <strong>Background on state-dependent diversification-rate estimation</strong>.</li>
</ul>

<p>These tutorials are for those interested in DEC models:</p>
<ul>
  <li>Provides a basic background for discrete biogeography models, with an emphasis on DEC: <a href="/tutorials/biogeo/biogeo_intro">Introduction to DEC models</a></li>
  <li>A simple time-constant DEC analysis on a fixed phylogeny within a 4-region system: <a href="/tutorials/biogeo/biogeo_simple">Simple DEC</a></li>
  <li>A paleogeography-informed DEC analysis on a fixed phylogeny within a 6-region system: <a href="/tutorials/biogeo/biogeo_epoch">Epoch DEC</a></li>
  <li>A paleogeography-informed DEC analysis on a jointly inferred phylogeny (topology, divergence times) within a 6-region system: <a href="/tutorials/biogeo/biogeo_dating">Biogeographic Dating with DEC</a></li>
</ul>

<p>These tutorials are for those interested in GeoSSE/FIG models:</p>
<ul>
  <li>A simple GeoSSE analysis using a 2-region system: <a href="/tutorials/geosse">Simple GeoSSE</a></li>
  <li>A time-constant FIG analysis in a 7-region system with multiple regional features:  <a href="/tutorials/multifig">MultiFIG</a></li>
  <li>A paleogeography-informed FIG analysis on a fixed phylogeny in a 7-region system with multiple regional features:  <a href="/tutorials/timefig_simple">TimeFIG</a></li>
  <li>A paleogeography-informed FIG analysis on a jointly inferred phylogeny (topology, divergence times) within a 7-region system with multiple regional features <a href="/tutorials/timefig_dating">Biogeographic Dating with TimeFIG</a></li>
</ul>
</blockquote>

      </div>
      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://github.com/revbayes/revbayes/discussions">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "default highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "default highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
