<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Tutorials</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/software">Software</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <h1 class="maintitle">

RevBayes Tutorials

</h1>
<hr>

<p>This list shows all of the RevBayes tutorials for learning various aspects of RevBayes and Bayesian phylogenetic analysis.
Each one explicitly walks you through model specification and analysis set-up for different phylogenetic methods.
These tutorials have been written for new users to learn RevBayes at home, at workshops, and in courses taught at the undergraduate and graduate levels. 
You may find that the styles are somewhat different between tutorials and that some  have overlapping content.</p>

<p>Please see the <a href="/tutorials/format">Tutorial Format</a> guide for details about how to read the tutorials.</p>

<p>Please see <a href="/tutorials/recommended">Recommended Software</a> for links to various software programs you may need to download in order to follow the tutorials.</p>

<p><a href="/developer/tutorial/" class="btn btn-warning" role="button">Contribute!</a></p>

<h3 id="intro">Introduction to RevBayes and MCMC</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Getting_Started/RB_Getting_Started.pdf"><p>Getting Started with RevBayes</p>
</a>

<p class="subtitle"><p>A very basic overview on how to use RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro/"><p>Rev Language Syntax</p>
</a>

<p class="subtitle"><p>A very short introduction to the Rev language</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/intro_graph_models/"><p>Introduction to Graphical Models</p>
</a>

<p class="subtitle"><p>A gentle introduction to graphical models, probabilistic programming, and MCMC using a simple linear regression example.</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MCMC_Binomial_Tutorial/RB_MCMC_Binomial_Tutorial.pdf"><p>Introduction to MCMC using RevBayes</p>
</a>

<p class="subtitle"><p>Introduction to MCMC Simulation using a simple Binomial Model</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/mcmc_archery/"><p>Introduction to MCMC using RevBayes</p>
</a>

<p class="subtitle"><p>A simple Archery example for building a hierarchical model and sampling under Markov Chain Monte Carlo</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Rev_Tutorial/RB_Rev_Tutorial.pdf"><p>Basic introduction to Rev &amp; MCMC</p>
</a>

<p class="subtitle"><p>General Rev language features and simple Poisson regression example</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MCMC_Tutorial/RB_MCMC_Tutorial.pdf"><p>Diagnosing MCMC performance</p>
</a>

<p class="subtitle"><p>How to assess the performance of MCMC simulations</p>
</p>
</div>



</div>

<h3 id="model_test">Model Selection and Testing</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_intro.html"><p>General Introduction to Model selection</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_subst_model.html"><p>Model selection of common substitution models for one locus</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>










<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_subst_model_rj.html"><p>Model averaging of substitution models</p>
</a>

<p class="subtitle"><p>Reversible-jump MCMC over substitution models</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_selection_bayes_factors/bf_partition_model.html"><p>Model selection of partition models</p>
</a>

<p class="subtitle"><p>Comparing relative model fit with Bayes factors</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_testing_pps/pps_data.html"><p>Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</p>
</a>

<p class="subtitle"><p>Model adequacy testing using posterior prediction (Data Version).</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/model_testing_pps/pps_inference.html"><p>Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</p>
</a>

<p class="subtitle"><p>Model adequacy testing using posterior prediction (Inference Version).</p>
</p>
</div>



</div>

<h3 id="phylo_basic">Standard tree inference</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/ctmc/"><p>Substitution Models</p>
</a>

<p class="subtitle"><p>Phylogenetic inference of nucleotide data using RevBayes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/selection/"><p>Inferring Selection through Codon Models</p>
</a>

<p class="subtitle"><p>Estimating dN/dS</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/partition/"><p>Partitioned data analysis</p>
</a>

<p class="subtitle"><p>Current Protocols in Bioninformatics - Phylogenetic Inference using RevBayes (Protocol #2)</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/clocks/"><p>Relaxed Clocks &amp; Time Trees</p>
</a>

<p class="subtitle"><p>Comparing relaxed clock models &amp; estimating rooted time trees</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph/"><p>Discrete morphology - Tree Inference</p>
</a>

<p class="subtitle"><p>Phylogenetic inference with discrete morphological data</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/morph/morph_more.html"><p>Discrete morphology - Ancestral State Estimation</p>
</a>

<p class="subtitle"><p>Ancestral State Estimation and Testing for Irreversibility</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_GeneConcatenation_Tutorial/RB_GeneConcatenation_Tutorial.pdf"><p>Alignment concatenation</p>
</a>

<p class="subtitle"><p>Species tree estimation from multiple concatenated gene sequence alignments</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_PhyloComparativeAdvanced_Tutorial/RB_PhyloComparativeAdvanced_Tutorial.pdf"><p>Phylogenetic Inference using RevBayes</p>
</a>

<p class="subtitle"><p>Continuous trait evolution with Ornstein-Uhlenbeck models</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MultispeciesCoalescent_Tutorial/RB_MultispeciesCoalescent_Tutorial.pdf"><p>Gene tree - species tree reconstruction</p>
</a>

<p class="subtitle"><p>Incomplete lineage sorting and the multispecies coalescent</p>
</p>
</div>



</div>

<h3 id="phylo_complex">Complex hierarchical models for phylogenetic inference</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/fbd/"><p>Combined-Evidence Analysis and the Fossilized Birth-Death Process for Stratigraphic Range Data</p>
</a>

<p class="subtitle"><p>Joint inference of divergence times and phylogenetic relationships of fossil and extant taxa</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/fbd/fbd_specimen.html"><p>Combined-Evidence Analysis and the Fossilized Birth-Death Process for Analysis of Fossil and Extant Specimens</p>
</a>

<p class="subtitle"><p>Joint inference of divergence times and phylogenetic relationships of fossil and extant taxa</p>
</p>
</div>



</div>

<h3 id="phylo_div_rate">Diversification Rate Estimation</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/divrate/div_rate_intro.html"><p>Introduction to Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Overview of Analyses, Models and Theory</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/simple.html"><p>Simple Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Comparing different constant-rate models of lineage diversification</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/ebd.html"><p>Episodic Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>Estimating Rate-Variation Through Time in Diversification Rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/env.html"><p>Environmental-dependent Speciation &amp; Extinction Rates</p>
</a>

<p class="subtitle"><p>Estimating Correlation between Diversification Rates and Environmental Characters</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/branch_specific.html"><p>Branch-Specific Diversification Rate Estimation</p>
</a>

<p class="subtitle"><p>How to estimate branch-specific shifts in diversification rates</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/bisse-intro.html"><p>Background on state-dependent diversification rate estimation</p>
</a>

<p class="subtitle"><p>An introduction to inference using state-dependent speciation and extinction (SSE) branching processes</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/bisse.html"><p>State-dependent diversification with BiSSE and MuSSE</p>
</a>

<p class="subtitle"><p>Inference using the binary/multiple state-dependent speciation and extinction (BiSSE/MuSSE) branching process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/divrate/sampling.html"><p>Diversification Rate Estimation with Missing Taxa</p>
</a>

<p class="subtitle"><p>How to estimate diversification rates with incomplete taxon sampling</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/sse/other-sse.html"><p>State-dependent diversification with HiSSE and ClaSSE</p>
</a>

<p class="subtitle"><p>Hidden state and cladogenetic state-dependent speciation and extinction (HiSSE &amp; ClaSSE) branching process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/fbd_range/"><p>Macroevolutionary Analysis of Stratigraphic Range Data</p>
</a>

<p class="subtitle"><p>Inference of diversification rates using the fossilized birth-death range process</p>
</p>
</div>



</div>

<h3 id="phylo_trait">Comparative methods</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_PhyloComparative_Tutorial/RB_PhyloComparative_Tutorial.pdf"><p>Continuous trait evolution and Brownian motion</p>
</a>

<p class="subtitle"><p>Phylogenetic analysis of continous characters and the comparative method</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/chromo/"><p>Chromosome Evolution</p>
</a>

<p class="subtitle"><p>Modeling chromosome evolution with ChromEvol, BiChroM, and ChromoSSE</p>
</p>
</div>



</div>

<h3 id="phylo_biogeo">Biogeography</h3>

<div class="tutorialbox">





<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_intro.html"><p>Introduction to Historical Biogeography</p>
</a>

<p class="subtitle"><p>Phylogeographic analysis under the dispersal-extinction-cladogenesis process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_simple.html"><p>Historical Biogeographic Analysis using a Simple DEC Model</p>
</a>

<p class="subtitle"><p>Ancestral range estimation of the silverswords under the dispersal-extinction-cladogenesis process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_epoch.html"><p>Historical Biogeographic Analysis using an epoch DEC Model</p>
</a>

<p class="subtitle"><p>Ancestral range estimation of the silverswords under the dispersal-extinction-cladogenesis process</p>
</p>
</div>







<div class="tutorial ">

<a class="title" href="/tutorials/biogeo/biogeo_dating.html"><p>Biogeographic dating analysis</p>
</a>

<p class="subtitle"><p>Biogeographic dating analysis under the dispersal-extinction-cladogenesis process</p>
</p>
</div>



</div>


      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "default highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "default highlighter-rouge";
//   }
//   
// });
</script>

<script src="/assets/js/base.js"></script>

<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>

  </body>
</html>
