<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
        <li><class="header-search">
  <form class="header-search-form" action="/search.html" method="get">
    <input type="text" id="search-box-nav" placeholder="Search..." name="query" style="border-color:black;background-color:white;height:32px;">
    <button type="submit" style="color:black; border-color:black;width:32px;height:32px;padding-top:4px">
        <img src="assets/img/search.png" height="22px" width="28px"/>
    </button>
  </form>
</div>

</li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Assessing Phylogenetic Reliability Using RevBayes and $P^{3}$</h1>
	<h3 class="subtitle">Model adequacy testing using posterior prediction (Data Version).</h3>
	<h4 class="authors">Lyndon M. Coghill, Will Freyman, Sebastian Höhna and Jeremy M. Brown</h4>
  <h5>Last modified on September 12, 2019</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/intro/">Getting Started with RevBayes and Rev Language Syntax</a></li>
          
            <li><a href="/tutorials/mcmc/">Introduction to Markov chain Monte Carlo (MCMC) Sampling</a></li>
          
            <li><a href="/tutorials/ctmc/">Nucleotide substitution models</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Data Files</strong>
        <ul id="data_files">
        
        
        
          <li><a href="/tutorials/model_testing_pps/data/primates_and_galeopterus_cytb.nex">primates_and_galeopterus_cytb.nex</a></li>
        
        </ul>
    
        
        <strong>Scripts</strong>
        <ul id="scripts">
        
        
        
          <li><a href="/tutorials/model_testing_pps/scripts/plot_results.R">plot_results.R</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_DataSummary.Rev">pps_DataSummary.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_GTR_Gamma_Inv_Model.Rev">pps_GTR_Gamma_Inv_Model.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_GTR_Inv_Model.Rev">pps_GTR_Inv_Model.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_GTR_Model.Rev">pps_GTR_Model.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_JC_Model.Rev">pps_JC_Model.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_MCMC_Simulation.Rev">pps_MCMC_Simulation.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_PValues.Rev">pps_PValues.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_Simulation.Rev">pps_Simulation.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_analysis_GTR.Rev">pps_analysis_GTR.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_analysis_JC.Rev">pps_analysis_JC.Rev</a></li>
        
          <li><a href="/tutorials/model_testing_pps/scripts/pps_tree_model.Rev">pps_tree_model.Rev</a></li>
        
        </ul>
    
</blockquote>


</div>
<h2 class="section" id="overview">Overview</h2>
<hr class="section" />

<p>This tutorial presents the general principles of assessing the
reliability of a phylogenetic inference through the use of posterior
predictive simulation (PPS). PPS works by assessing the fit of an
evolutionary model to a given dataset, and analyzing several test
statistics using a traditional goodness-of-fit framework to help explain
where a model may be the most inadequate.</p>

<h2 class="section" id="introduction">Introduction</h2>
<hr class="section" />

<p>Assessing the fit of an evolutionary model to the data is critical as
using a model with poor fit can lead to spurious conclusions. However, a
critical evaluation of absolute model fit is rare in evolutionary
studies. Posterior prediction is a Bayesian approach to assess the fit
of a model to a given dataset
<a class="citation" href="#Bollback2002">(Bollback 2002; Brown 2014; Höhna et al. 2018)</a>, that relies on the
use of the posterior and the posterior predictive distributions. The
posterior distribution is the standard output from Bayeisan phylogenetic
inference. The posterior predictive distribution represents a range of
possible outcomes given the assumptions of the model. The most common
method to generate these possible outcomes, is to sample parameters from
the posterior distribution, and use them to simulate new replicate
datasets (<a href="#pps_schematic"></a>). If these simulated datasets differ from the empirical
dataset in a meaningful way, the model is failing to capture some
salient feature of the evolutionary process.</p>

<figure id="pps_schematic"><p><img src="figures/pps.png" /></p>
<figcaption>A schematic presentation of data- versus
inference-based approaches to assessing model plausibility with
posterior predictivesimulation. Most statistics proposed for testing
model plausibility compare data-based characteristics of the original
data set to the posterior predictive data sets (e.g., variation in
GC-content across species). ‘RevBayes‘ additionally implements test
statistics that compare the inferences resulting from different data
sets (e.g., the distribution of posterior probability across
topologies). Multiple sequence alignments (MSAs) are represented as
shaded matrices and arrows originating from MSAs point to the MCMC
samples of tree topologies and scalar model parameters ($\theta$)
resulting from Bayesian analysis of that MSA. Subscripts of MCMC samples
taken during analysis of the original data index the samples
$(1,    …, n)$. Subscripts for each posterior predictive data set
indicate which MCMC sample was used in its simulation. Subscripts for
MCMC samples resulting from analysis of a posterior predictive data set
first indicate the posterior predictive data set that was analyzed and
next index the MCMC samples from analysis of that particular data set
$(1, …, m)$.</figcaption>
</figure>

<p>The framework to construct posterior predictive distributions, and
compare them to the posterior distribution is conveniently built in to
‘RevBayes‘ <a class="citation" href="#Hoehna2014b">(Höhna et al. 2014; Höhna et al. 2016; Höhna et al. 2018)</a>. In this tutorial we will walk you through using this
functionality to perform a complete posterior predictive simulation on
an example dataset.</p>

<h3 class="subsection" id="data--versus-inference-based-comparisons">Data- Versus Inference-Based Comparisons</h3>
<hr class="subsection" />

<p>Most statistics proposed for testing model plausibility compare
data-based characteristics of the original data set to the posterior
predictive data sets (e.g., variation in GC-content across species). In
data-based assessments of model fit one compares the empirical data to
data simulated from samples of the posterior distribution. ‘RevBayes‘
additionally implements test statistics that compare the inferences
resulting from different data sets (e.g., the distribution of posterior
probability across topologies). These are called inference-based
assessments of model fit. For these assessments one must run an MCMC
analysis on each simulated data set, and then compare the inferences
made from the simulated data to the inference made from the empirical
data.</p>

<p>In this tutorial we will only cover the data-based method of assessing model plausibility. 
The inference-based method can be a powerful tool that you may want to explore at another
time.</p>

<h3 class="subsection" id="substitution-models">Substitution Models</h3>
<hr class="subsection" />

<p>The models we use here are equivalent to the models described in the
previous exercise on substitution models (continuous time Markov
models). To specify the model please consult the previous exercise.
Specifically, you will need to specify the following substitution
models:</p>

<ul>
  <li>Jukes-Cantor (JC) substitution model (missing reference)</li>
  <li>General-Time-Reversible (GTR) substitution model (missing reference)</li>
  <li>Gamma (+G) model for among-site rate variation (missing reference)</li>
  <li>Invariable-sites (+I) model (missing reference)</li>
</ul>

<h2 class="section" id="assessing-model-fit-with-posterior-prediction">Assessing Model Fit with Posterior Prediction</h2>
<hr class="section" />

<p>The entire process of posterior prediction can be executed by using the
<strong>pps_analysis_JC.Rev</strong> script in the <strong>scripts</strong> folder. If you
were to type the following command into ‘RevBayes‘:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    source("scripts/pps_analysis_JC.Rev")
</code></pre></div></div>
<p>the entire data-based posterior prediction process would run on the
example dataset. However, in this tutorial, we will walk through each
step of this process on an example dataset to explain what is happening in detail.</p>

<h3 class="subsection" id="empirical-mcmc-analysis">Empirical MCMC Analysis</h3>
<hr class="subsection" />

<p>To begin, we first need to generate a posterior distribution from which
to sample for simulation. This is the normal, and often only, step
conducted in phylogenetic studies. Here we will specify our dataset,
evolutionary model, and run a traditional MCMC analysis.</p>

<p>This code is all in the <strong>data_pp_analysis_JC.Rev</strong> file.</p>

<h4 class="subsubsection" id="set-up-the-workspace">Set up the workspace</h4>
<hr class="subsubsection" />

<p>First, let’s set up some workspace variables we’ll need. We’ll specify a
general name to apply to your analysis. This will be used for future
output files, so make sure it’s something clear and easy to understand.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    analysis_name = "pps_example"
    model_name = "JC"
    model_file_name = "scripts/pps_"+model_name+"_Model.Rev"
</code></pre></div></div>

<p>Now specify and read in the input file. This is your sequence alignment
in NEXUS format.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    inFile = "data/primates_and_galeopterus_cytb.nex"
    data &lt;- readDiscreteCharacterData(inFile)
</code></pre></div></div>

<h4 class="subsubsection" id="specify-the-model">Specify the model</h4>
<hr class="subsubsection" />

<p>Now we’ll call a separate script <strong>pps_JC_Model.Rev</strong> that specifies the JC
model:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    source( model_file_name ) 
</code></pre></div></div>
<p>If we open the <strong>pps_JC_Model.Rev</strong> script, most of this script should look
familiar from the <a href="/tutorials/ctmc/">Nucleotide substitution models</a>. Nevertheless, we will look at some parts 
of script for a very brief repitition.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    Q &lt;- fnJC(4)
</code></pre></div></div>
<p>Here we are specifying that we should use the Jukes-Cantor model, and
have it applied uniformly to all sites in the dataset. While this
obviously is not likely to be a good fitting model for most datasets, we
are using it for simplicity of illustrating the process.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    topology ~ dnUniformTopology(names)
</code></pre></div></div>
<p>This sets a uniform prior on the tree topology.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    br_lens[i] ~ dnExponential(10.0)
</code></pre></div></div>
<p>This sets an exponential distribution as the branch length prior.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    phylogeny := treeAssembly(topology, br_lens)
</code></pre></div></div>
<p>This builds the tree by combining the topology with branch length
support values.</p>

<h4 class="subsubsection" id="run-the-mcmc">Run the MCMC</h4>
<hr class="subsubsection" />

<p>Now let’s run MCMC on our empirical dataset, just like a normal
phylogenetic analysis. Here we use the Jukes-Cantor substition model.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    source("scripts/pps_MCMC_Simulation.Rev")
</code></pre></div></div>
<p>After specifying a model, we will conduct the MCMC analysis with the
script <strong>pps_MCMC_Simulation.Rev</strong>. Much of that script should look familiar
from the <a href="/tutorials/ctmc/">Nucleotide substitution models</a>, but let’s take a look at the script and
break down and revisit some of the major pieces as a refresher. Much of
this will be a familiar to you from the <a href="/tutorials/ctmc/">Nucleotide substitution models</a>.</p>

<p>First we need to specify some monitors.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    mni = 0
    monitors[++mni] = mnModel(filename="output" + n + "/" + analysis_name + "_posterior.log",printgen=10, separator = TAB)
    monitors[++mni] = mnFile(filename="output" + n + "/" + analysis_name + "_posterior.trees",printgen=10, separator = TAB, phylogeny)
    monitors[++mni] = mnScreen(printgen=1000, TL)
    monitors[++mni] = mnStochasticVariable(filename="output" + n + "/" + analysis_name + "_posterior.var",printgen=10)
</code></pre></div></div>
<p>Here the monitors are just being advanced by the value of mni which is
being incremented each generation.</p>

<p>Next, we will call the MCMC function, passing it the model, monitors and
moves we specified above to set to build our mymcmc object.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    mymcmc = mcmc(mymodel, monitors, moves, nruns=2, combine="mixed")
</code></pre></div></div>
<p>Specify a burnin for this run.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    mymcmc.burnin(generations=2000,tuningInterval=200)
</code></pre></div></div>
<p>Finally, we will execute the run for a specified number of generations.
The number of generations and printed generations is important to consider here for a
variety of reasons, in particular for posterior predictive simulation.
When we simulate datasets in the next step, we can only simulate 1
dataset per sample in our posterior. So, while the number of posterior
samples will almost always be larger than the number of datasets we will
want to simulate, it’s something to keep in mind.</p>

<p>Now let’s actually run the MCMC in ‘RevBayes‘.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    mymcmc.run(generations=10000)
</code></pre></div></div>

<h4 class="subsubsection" id="mcmc-output">MCMC output</h4>
<hr class="subsubsection" />

<p>After the process completes, the results can be found in the
<strong>output_JC</strong> folder. You should see a number of familiar looking
files, all with the name we provided under the <strong>analysis_name</strong>
variable, <strong>pps_example</strong> in this case. Since we set the number of runs
(nruns=2) in our MCMC, there will be two files of each type (.log .trees
.var) with an _<em>N</em> where <em>N</em> is the run number. You will also see 3
files without any number in their name. These are the combined files of
the output. These will be the files we use for the rest of the process.
If you open up one of the combined .var file, you should see that there
are 1000 samples. This was produced by our number of generations
(10,000) divided by our number of printed generations (10), which we
specified earlier. This is important to note, as we will need to thin
these samples appropriately in the next step to get the proper number of
simulated datasets.</p>

<h2 class="section" id="posterior-predictive-data-simulation">Posterior Predictive Data Simulation</h2>
<hr class="section" />

<p>The next step of posterior predictive simulation is to simulate new
datasets by drawing samples and parameters from the posterior
distribution generated from the empirical MCMC anlaysis.</p>

<p>In the <strong>pps_analysis_JC.Rev</strong> script, that is conducted using
the following lines of RevScript:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    source("scripts/pps_Simulation.Rev")
</code></pre></div></div>
<p>Let’s take a look at each of the calls in the
<strong>pps_Simulation.Rev</strong> script so that we can understand
what is occurring.</p>

<p>First, we read in the trace file of the Posterior Distribution of
Variables.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    trace = readStochasticVariableTrace("output_" + model_name + "/" + analysis_name + "_posterior.var", delimiter=TAB)
</code></pre></div></div>
<p>Now we call the <strong>posteriorPredictiveSimulation()</strong> function, which
accepts any valid model of sequence evolution, and output directory, and
a trace. For each line in the trace, it will simulate a new dataset
under the specified model.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    pps = posteriorPredictiveSimulation(mymodel, directory="output" + "/" + analysis_name + "_post_sims", trace)
</code></pre></div></div>
<p>Now we run the posterior predictive simulation, generating a new dataset
for each line in the trace file that was read in. This is the part where
we need to decide how many simulated datasets we want to generate. If we
just use the pps.run() command, one dataset will be generated for each
sample in our posterior distribution. In this case, since we are reading
in the combined posterior trace file with 2000 samples, it would generate
2000 simulated datasets. If you want to generate fewer, say 1000 datasets,
you need to use the thinning argument as above. In this case, we are
thinning the output by 2, that is we are dividing our number of samples
by 2. So that in our example case, we will end up simulating 1000 new
datasets.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    pps.run(thinning=2)
</code></pre></div></div>
<p>This process should finish in just a minute or two. If you look in the
<strong>output_JC</strong> folder, there should be another folder called
<strong>pps_example_post_sims</strong>. This folder is where the simulated
datasets are saved. If you open it up, you should see 100 folders named
posterior_predictive_sim_<em>N</em>. Where <em>N</em> is the number of the
simulated dataset. In each of these folders, you should find a seq.nex
file. If you open one of those files, you’ll see it’s just a basic NEXUS
file. These will be the datasets we analyze in the next step.</p>

<h2 class="section" id="calculating-the-test-statistics">Calculating the Test Statistics</h2>
<hr class="section" />

<p>Now we will calculate the test statistics from the empirical data and
the simulated data sets. The part in the <strong>pps_analysis_JC.Rev</strong>
script that generates the test statistics is the following line:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    num_post_sims = listFiles(path="output_"+model_name+"/" + analysis_name + "_post_sims").size()
    source("scripts/pps_DataSummary.Rev")
</code></pre></div></div>
<p>We will look at the major concepts of the
<strong>pps_DataSummary.Rev</strong> to better understand how it
works. For a more complete discussion of the statistics involved, please
review <a class="citation" href="#Hoehna2018a">(Höhna et al. 2018)</a>. In general, this script and these
statistics work by calculating the statistics of interest across each
posterior distribution from the simulated datasets, and comparing those
values to the values from the empirical posterior distribution.</p>

<p>The current version of this script generates over 30 summary statistics
including:</p>

<ul>
  <li>Number Invariant Sites</li>
  <li>Max Pairwise Difference</li>
  <li>Max GC Content</li>
  <li>Min GC Content</li>
  <li>Mean GC Content</li>
</ul>

<p>There are functions built-in to ‘RevBayes‘ to calculate these values for
you. Here are some examples from the
<strong>pps_DataSummary.Rev</strong> script:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    max_pd          = sim_data.maxPairwiseDifference( excludeAmbiguous=FALSE )
    mean_gc_1       = sim_data.meanGcContentByCodonPosition(1, excludeAmbiguous=FALSE )
    var_gc_1        = sim_data.varGcContentByCodonPosition(1, excludeAmbiguous=FALSE )
</code></pre></div></div>
<p>These same statistics are calculated for both the posterior
distributions from the simulated datasets and the posterior distribution
from the empirical dataset.</p>

<h2 class="section" id="calculating-p-values-and-effect-sizes">Calculating <em>P</em>-values and effect sizes</h2>
<hr class="section" />

<p>Once we have the test statistics calculated for the simulated and
empirical posterior distributions, we can compare the simulated to the
empirical to get a goodness-of-fit. One simple way to do this is to
calculate a posterior predictive <em>P</em>value for each of the test
statistics of interest. This is done in the
<strong>data_pp_analysis_JC.Rev</strong> with the following lines :</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    emp_pps_file = "results_" + model_name + "/empirical_data_" + analysis_name + ".csv"
    sim_pps_file = "results_" + model_name + "/simulated_data_" + analysis_name + ".csv"
    outfileName = "results_" + model_name + "/data_pvalues_effectsizes_" + analysis_name + ".csv"
    source("scripts/pps_PValues.Rev")
</code></pre></div></div>
<p>The 4 posterior predictive <em>P</em>-values currently calculated by ‘RevBayes‘
are:</p>

<ul>
  <li>Lower 1-tailed</li>
  <li>Upper 1-tailed</li>
  <li>2-tailed</li>
  <li>Midpoint</li>
</ul>

<p>The posterior predictive <em>P</em>-value for a lower one-tailed test is the
proportion of samples in the distribution where the value is less than
or equal to the observed value, calculated as:</p>

\[p_l=p(T(X_{rep})\leqslant T(X_{emp}))\]

<p>The posterior predictive <em>P</em>-value for a lower one-tailed test is the
proportion of samples in the distribution where the value is greater
than or equal to the observed value, calculated as:</p>

\[p_u=p(T(X_{rep})\geqslant T(X_{emp}))\]

<p>and the two-tailed posterior predictive <em>P</em>-value is simply twice the
minimum of the corresponding one-tailed tests calculated as:</p>

\[p_2=2min(p_l,p_u)\]

<p>Finally, the midpoint posterior predictive <em>P</em>-value iscalculated as:</p>

\[p_m=p(T(X_{rep})\leqslant T(X_{emp})) + \frac{1}{2} p(T(X_{rep})= T(X_{emp}))\]

<p>For more a more detailed description please look at <a class="citation" href="#Hoehna2018a">(Höhna et al. 2018)</a>.</p>

<h3 class="subsection" id="the-pps_pvaluesrev-script">The pps_PValues.Rev script</h3>
<hr class="subsection" />

<p>Let’s take a look at the <strong>pps_PValues.Rev</strong> script to
get a better idea of what is happening.</p>

<p>Using the equations outlined above, this script reads in the simulated
and empirical data we just calculated, and calls a few simple functions
for each of the test statistics:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    ## Calculate and return a vector of lower, equal, and upper pvalues for a given test statistic
    p_values &lt;- posteriorPredictiveProbability(numbers, empValue)

    ## 1-tailed
    lower_p_value &lt;- p_values[1] 
    equal_p_value &lt;- p_values[2] 
    upper_p_value &lt;- p_values[3] 

    ## mid-point
    midpoint_p_value = lower_p_value + 0.5*equal_p_value

    ## 2-tailed
    two_tail_p_value = 2 * (min(v(lower_p_value, upper_p_value)))
</code></pre></div></div>
<p>these snippets calculate the various p-values of interest to serve as
our goodness-of-fit test.</p>

<h3 class="subsection" id="posterior-predictive-effect-size">Posterior predictive effect size</h3>
<hr class="subsection" />

<p>Another way that you can calculate the magnitude of the discrepancy
between the empirical and PP datasets is by calculating the effect size
of each test statistic. Effect sizes are useful in quantifying the
magnitude of the difference between the empirical value and the
distribution of posterior predictive values. The test statistic effect
size can be calculated by taking the absolute value of the difference
between the median posterior predictive value and the empirical value
divided by the standard deviation of the posterior predictive
distribution (missing reference). Effect sizes are calculated automatically
for the inference based test statistics in the P analysis. The effect
sizes for each test statistics are stored in the same output file as the
<em>P</em>-values.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    effect_size = abs((m - empValue) / stdev(numbers))
</code></pre></div></div>
<p>calculates the effect size of a given test statistic.</p>

<h2 class="section" id="viewing-the-results">Viewing the Results</h2>
<hr class="section" />

<p>Once execution of the script is complete, you should see a new
directory, named <strong>results_JC</strong>. In this folder there should be 3
files. Each of these is a simple comma-delimited (csv) file containing the
test statistic calculation output.</p>

<ul>
  <li>empirical_pps_example.csv</li>
  <li>pps_example.csv</li>
  <li>pvalues_pps_example.csv</li>
</ul>

<p>If you have these 3 files, and there are results in them, you can go
ahead and quit ‘RevBayes‘.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    q()
</code></pre></div></div>

<figure id="pps_mean_gc"><p><img src="figures/mean_gc.png" width="50%" /></p>
<figcaption>The distribution of mean GC values calculated
from the simulated data set is shown. The dotted line represents the
mean GC calculated from the empirical data set.The Jukes-Cantor model
does not adequately describe the empirical data. This plot was generated
in <code class="language-plaintext highlighter-rouge">R</code>using the <strong>scripts/plot_results.R</strong> script.</figcaption>
</figure>

<p>You can get an estimate of how the model performed by examining the
<em>P</em>-values in the  <strong>data_pvalues_effectsizes_pps_example.csv</strong> file. In this
example, a quick view shows us that most of the statistics show a value
less than 0.05. This leads to a rejection of the model, suggesting that
the model employed is not a good fit for the data. This is the result we
expected given we chose a very simplistic model (JC) that would likely
be a poor fit for our data. However, it’s a good example of the
sensitivity of this method, showing how relatively short runtimes and a
low number of generations will still detect poor fit.</p>

<p>You can also use <code class="language-plaintext highlighter-rouge">R</code>to plot the results. Run the
<code class="language-plaintext highlighter-rouge">R</code>script <strong>scripts/plot_results.R</strong>. This will generate a
pdf plot for every test statistic. See <a href="#pps_mean_gc"></a> and its legend to learn how to interpret these plots.</p>

<h2 class="section" id="exercises">Exercises</h2>
<hr class="section" />

<p>Included in the <strong>scripts</strong> folder is a second model script called
<strong>pps_GTR_Model.Rev</strong>. As a personal exercise and a good test case, take
some time now, and run the same analysis, substituting the
<strong>pps_GTR_Model.Rev</strong> model script for the <strong>pps_JC_Model.Rev</strong> script we used
in the earlier example. You should get different results, this is an
excellent chance to explore the results and think about what they
suggest about the fit of the specified model to the dataset.</p>

<h3 id="additional-exercises">Additional Exercises</h3>

<p><strong>Some Questions to Keep in Mind:</strong></p>

<ul>
  <li>
    <p>Do you find the goodness-of-fit results to suggest that the GTR or
JC model is a better fit for our data?</p>
  </li>
  <li>
    <p>Which test statistics seem to show the strongest effect from the use
of a poorly fitting model?</p>
  </li>
</ul>

<blockquote class="aside"><h2>Batch Processing of Large Datasets</h2><p>In this tutorial you have learned how to use ‘RevBayes‘ to assess the
fit of a substitution model to a given sequence alignment. As you have
discovered, the observed data should be plausible under the posterior
predictive simulation if the model is reasonable. In phylogenetic
analyses we choose a model, which explicitly assumes that it provides an
reasonable explanation of the evolutionary process that generated our
data. However, just because a model may be the ’best’ model available,
does not mean it is an appropriate model for the data. This distinction
becomes both more critical and less obvious in modern analyses, where
the number of genes often number in the thousands. Posterior predictive
simulation in ‘RevBayes‘, allows you to easily check model fit for a
large number of genes by using global summaries to check the posterior
predictive distributions with a comfortable goodness-of-fit style
framework.</p>

<p>The process described above is for a single gene or alignment. However,
batch processing a large number of genes with this method is a
relatively straight forward process.</p>

<p>‘RevBayes‘ has built in support for MPI so running ‘RevBayes‘ on more
than a single processor, or on a cluster is as easy as calling it with
openmpi.</p>

<p>For example:</p>
<pre><code class="language-{:bash}">    mpirun -np 16 rb-mpi scripts/full_analysis.Rev
</code></pre>
<p>would run the entire posterior predictive simulation analysis on a
single dataset using 16 processors instead of a single processor. Use of
the MPI version of ‘RevBayes‘ will speed up the process dramatically.</p>

<p>Setting up the <strong>full_analysis.Rev</strong> script to cycle through a large
number of alignments is relatively simple as well. One easy way is to
provide a list of the data file names, and to loop through them. As an
example:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    data_file_list = "data_file_list.txt" 
    data_file_list &lt;- readDiscreteCharacterData(data_file_list)
    file_count = 0

    for (n in 1:data_file_list.size()) {

    FULL_ANALYSIS SCRIPT GOES HERE

    file_count = file_count + 1
    }
</code></pre></div></div>
<p>Then, anywhere in the full analysis portion of the script that the
lines</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    inFile = "data/8taxa_500chars_GTR.nex"
    analysis_name = "pps_example"
</code></pre></div></div>
<p>appear, you would replace them with something along the lines of:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    inFile = n
    analysis_name = "pps_example_" + file_count
</code></pre></div></div>
<p>This should loop through all of the data files in the list provided, and
run the full posterior predictive simulation analysis on each file.
Using a method like this, and combining it with the MPI call above, you
can scale this process up to multiple genes and spread the computational
time across several cores to speed it up.</p>
</blockquote>


<ol class="bibliography"><li><span id="Bollback2002">Bollback J.P. 2002. Bayesian model adequacy and choice in phylogenetics. Molecular Biology and Evolution. 19:1171–1180.</span>

<a href="https://doi.org/10.1093/oxfordjournals.molbev.a004175">10.1093/oxfordjournals.molbev.a004175</a>

</li>
<li><span id="Brown2014">Brown J.M. 2014. Predictive approaches to assessing the fit of evolutionary models. Systematic Biology. 63:289–292.</span>

<a href="https://doi.org/10.1093/sysbio/syu009">10.1093/sysbio/syu009</a>

</li>
<li><span id="Hoehna2016b">Höhna S., Landis M.J., Heath T.A., Boussau B., Lartillot N., Moore B.R., Huelsenbeck J.P., Ronquist F. 2016. RevBayes: Bayesian Phylogenetic Inference Using Graphical Models and an Interactive Model-Specification Language. Systematic Biology. 65:726–736.</span>

<a href="https://doi.org/10.1093/sysbio/syw021">10.1093/sysbio/syw021</a>

</li>
<li><span id="Hoehna2018a">Höhna S., Coghill L.M., Mount G.G., Thomson R.C., Brown J.M. 2018. P^3: Phylogenetic Posterior Prediction in RevBayes. Molecular Biology and Evolution. 35:1028–1034.</span>

<a href="https://doi.org/10.1093/molbev/msx286">10.1093/molbev/msx286</a>

</li>
<li><span id="Hoehna2014b">Höhna S., Heath T.A., Boussau B., Landis M.J., Ronquist F., Huelsenbeck J.P. 2014. Probabilistic Graphical Model Representation in Phylogenetics. Systematic Biology. 63:753–771.</span>

<a href="https://doi.org/10.1093/sysbio/syu039">10.1093/sysbio/syu039</a>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
