<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: PhD position&#58; Developing new methods and models for robust gene-tree estimation</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <h1>PhD position&#58; Developing new methods and models for robust gene-tree estimation</h1>

<b>Location:</b> GeoBio-Center, Ludwig-Maximilians-Universität, München, Germany <br>

<b>PI:</b> <a href="mailto:phylomatics@gmail.com?subject=PhD-Position">Sebastian Höhna</a> <br>

<b>Application Deadline:</b> 10/31, 2018<br>

<b>Expected Startdate:</b> 1/1, 2019<br>

<b>Length of contract:</b> 3 years <br>

<b>Requirements:</b> Master’s degree in evolutionary biology, computer science, mathematics, statistics, or a related field. <br>

<b>Salary:</b> Standard DFG salary scheme (TVL-E13) <br>

<h2>Description</h2>

<p>I invite applications for a doctoral position to develop new methods and models for robust gene-tree estimation in my research group at the GeoBio-Center of the Ludwig-Maximilians-Universität (LMU), München. The position is funded by the DFG Emmy Noether program, and is available for 3 years (according to German funding regulations). The position is full-time and research only (no classes and teaching required). The position should start on 1 January 2019 or as soon as possible thereafter.</p><p>My group is broadly working on theory and computational methods for Bayesian inference of phylogeny (<a href="http://www.evol.bio.lmu.de/research/hoehna">http://www.evol.bio.lmu.de/research/hoehna</a>). The research directions include phylogeny inference, divergence time estimation, diversification rate estimation and model testing. All of our methods are implemented in the open-source program <a href="http://www.RevBayes.com">RevBayes</a> which is the successor software of the popular program MrBayes. The successful applicant will be part of our vibrant RevBayes group and will contribute to further development of the program. There will be opportunities for the successful applicant to work with and visit the research groups of my collaborators in Europe and the USA. Furthermore, I expect the candidate to become actively involved in our RevBayes workshops and hackathons.</p><p>I have recently been awarded an Emmy Noether grant from the DFG (German Science Foundation) which will fund at least 3 positions over the next 5 years. This advertisement is for one of these positions and the applicant will join a young, dynamic and rapidly growing group. My group will be moving to the GeoBio-Center of the LMU Munich, one of Europe’s top Universities (&#35;32 world-wide&#59; &#35;8 in Europe&#59; &#35;1 in Germany&#59; <a href="https://www.timeshighereducation.com/world-university-rankings/lmu-munich">Ranking</a>). The GeoBio-Center is located at the Königsplatz which is in walking distance to the historic city center (Marienplatz) and English Garden (city park with 3.75 km² area). The GeoBio-Center is highly interdisciplinary and consists of researchers from different departments including paleontology, molecular and evolutionary biology, zoology and botany.</p><p>The main research topic for the PhD project is robust estimation of gene trees. Today we have several databases with whole genomes which we would like to use to build phylogenetic trees. However, different genes have different evolutionary histories. To be able to understand why gene trees are discordant, we have to be able to estimate gene trees correctly in the first place. Thus, we need to develop realistic models of the substitution process for each gene. For example, we need to develop and test time-reversible and non-reversible substitution processes, lineage-heterogeneous substitution processes, etc. The foundation of these models is already implemented in RevBayes. The PhD student will apply and explore different substitution models and, depending on the results, develop the next steps for robust gene tree inference.</p><p>Applicants should have a Master’s degree, completed or completion imminent, in evolutionary biology, computer science, mathematics, statistics, or a related field.  Some knowledge and experience in programming (C++, Java, Python or R), phylogenetic inference as well as Bayesian statistics is beneficial. Training in these skills will be provided depending on need. The thesis will be written in English. No knowledge of German is required but some basic knowledge will be helpful outside of work. Enthusiasm, determination and the capacity to work independently are essential. Own ideas complementing the current research direction are highly appreciated.</p><p>The position will be compensated according to the standard DFG salary scheme (TVL-E13). The salary is very competitive and includes benefits such as health care, pension, unemployment insurance and child support (if applicable).</p><p>Further information can be found at <a href="http://www.evol.bio.lmu.de/research/hoehna">http://www.evol.bio.lmu.de/research/hoehna</a>, and questions should be directed to Sebastian Höhna (phylomatics@gmail.com). Applications, including a current CV, letter of motivation (1 page) and names and contact details of two referees should be sent to Sebastian Höhna before the deadline of 31 October 2018.</p>








      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "default highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "default highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
