<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="http://localhost:4000/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Postdoc position&#58; Developing methods and models for gene-tree species-tree estimation</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/software">Software</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="https://revbayes.github.io/revscripter/">RevScripter</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <h1>Postdoc position&#58; Developing methods and models for gene-tree species-tree estimation</h1>

<b>Location:</b> GeoBio-Center, Ludwig-Maximilians-Universität, München, Germany <br>

<b>PI:</b> <a href="mailto:phylomatics@gmail.com?subject=Postdoc-Position">Sebastian Höhna</a> <br>

<b>Application Deadline:</b> 10/31, 2018<br>

<b>Expected Startdate:</b> 1/1, 2019<br>

<b>Length of contract:</b> 2 years, with a further 2 years of funding available depending on progress and interests. <br>

<b>Requirements:</b> PhD degree in evolutionary biology, computer science, mathematics, statistics, or a related field. <br>

<b>Salary:</b> Standard DFG salary scheme (TVL-E13 to E14 based on prior experience) <br>

<h2>Description</h2>

<p>I invite applications for a postdoctoral research position to develop new methods and models for gene-tree species-tree estimation in my research group at the GeoBio-Center of the Ludwig-Maximilians-Universität (LMU), München. The position is funded by the DFG Emmy Noether program, and is initially available for 2 years, with a further 2 years of funding available depending on progress and interests. The position should start on 1 January 2019 or as soon as possible thereafter.</p><p>My group is broadly working on theory and computational methods for Bayesian inference of phylogeny (<a href="http://www.evol.bio.lmu.de/research/hoehna">http://www.evol.bio.lmu.de/research/hoehna</a>). The research directions include phylogeny inference, divergence time estimation, diversification rate estimation and model testing. All of our methods are implemented in the open-source program <a href="http://www.RevBayes.com">RevBayes</a> which is the successor software of the popular program MrBayes. The successful applicant will be part of our vibrant RevBayes group and will contribute to further development of the program. There will be opportunities for the successful applicant to work with and visit the research groups of my collaborators in Europe and the USA. Furthermore, I expect the candidate to become actively involved in our RevBayes workshops and hackathons.</p><p>I have recently been awarded an Emmy Noether grant from the DFG (German Science Foundation) which will fund at least 3 positions over the next 5 years. This advertisement is for one of these positions and the applicant will join a young, dynamic and rapidly growing group. My group will be moving to the GeoBio-Center of the LMU Munich, one of Europe’s top Universities (&#35;32 world-wide&#59; &#35;8 in Europe&#59; &#35;1 in Germany&#59; <a href="https://www.timeshighereducation.com/world-university-rankings/lmu-munich">Ranking</a>). The GeoBio-Center is located at the Königsplatz which is in walking distance to the historic city center (Marienplatz) and English Garden (city park with 3.75 km² area). The GeoBio-Center is highly interdisciplinary and consists of researchers from different departments including paleontology, molecular and evolutionary biology, zoology and botany.</p><p>The main question the postdoctoral researcher will work on is how to model biological processes that explain gene tree incongruence by developing models for gene-tree species-tree inference using genomic data. The process should model incomplete lineage sorting due to the multi-species coalescent, migration/hybridization between closely related species, and gene duplication and loss events. The models should be implemented in our software RevBayes using C++ for computational efficiency. Additionally, the postdoctoral researcher should perform simulation studies of the new models, apply them to empirical data, and lead the writing of manuscripts describing the work.</p><p>Applicants must have completed their PhD by the position start date, with degrees in in evolutionary biology, computer science, mathematics, statistics, or a related field. The candidate must be enthusiastic and capable of working independently. Additionally, proficiency in programming and C++ and experience in phylogenetic methods development as well as Bayesian statistics are preferred skills. There will also be opportunities for the postdoc to develop their own, independent research ideas that are complementary to the goals of this project.</p><p>The position will be compensated according to the standard DFG salary scheme (TVL-E13 to E14 based on prior experience). The salary is very competitive and includes benefits such as health care, pension, unemployment insurance and child support (if applicable).</p><p>Further information can be found at <a href="http://www.evol.bio.lmu.de/research/hoehna">http://www.evol.bio.lmu.de/research/hoehna</a>, and questions should be directed to Sebastian Höhna (phylomatics@gmail.com). Applications, including a current CV, letter of motivation (1 page) and names and contact details of two referees should be sent to Sebastian Höhna before the deadline of 31 October 2018.</p>








      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "default highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "default highlighter-rouge";
//   }
//   
// });
</script>

<script src="/assets/js/base.js"></script>

<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>

  </body>
</html>
