<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Molecular dating</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="/interfaces">Interfaces</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Molecular dating</h1>
	<h3 class="subtitle">Estimating speciation times using the fossilized birth-death range process</h3>
	<h4 class="authors">Rachel Warnock, Sebastian Höhna, Tracy Heath, April  Wright and Walker Pett</h4>
  <h5>Last modified on September 17, 2019</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/dating/nodedate.html">Molecular dating</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Other</strong>
        <ul id="other_files">
        
        
        
          <li><a href="/tutorials/dating/scripts/tree_FBD.Rev">tree_FBD.Rev</a></li>
        
        </ul>
    
</blockquote>


</div>
<h1 class="section" id="exercise-4">Exercise 4</h1>

<p>In <a href="/tutorials/dating/nodedate">exercise 3</a> we inferred the phylogeny of extant bears and a timeline for their evolution using node dating. Node dating has several caveats, one which is that we hardly used any of the available fossil information.</p>

<p>In this exercise we will explictly incorporate information from the fossil record into the diversification process (i.e. the tree model). The <em>fossilized birth-death (FBD) process</em> is a model that assumes extant samples and fossils were generated by the same underlying evolutionary process <a class="citation" href="#Stadler2010">(Stadler 2010; Heath et al. 2014)</a>. The model is an extension of the birth-death process that incorporates the fossil recovery process, meaning extinct samples are directly incorporated as part of the tree and may be sampled along internal branches. This is in contrast to node dating, where we only used information from the fossil record to constrain the node ages but the fossils were not assummed to be part of the tree.
If a fossil falls directly along branch that has descendant samples in tree, it is referred to as a <em>sampled ancestor</em>.</p>

<p>The data we have available for the fossil bear species are <em>stratigraphic ranges</em>. This means that for each species we have dates for the first and last appearances, instead of fossil occurrence data. 
We will therefore use the fossilized birth-death <em>range</em> process <a class="citation" href="#Stadler2018">(Stadler et al. 2018)</a>, rather than the original FBD model, which is more suitable for specimen level datasets.</p>

<figure id="fig_fbd_gm"><p><img src="figures/tikz/fbd_gm.png" width="500" /></p>
<figcaption>A graphical model of the fossilized birth-death model describing the 
generation of the time tree (in
<a href="#fig_module_gm"></a>) used in this tutorial. The parameters of the
fossilized birth-death process are labeled in orange. The speciation,
extinction and fossilization rates are stochastic nodes (circles) drawn
from exponential distributions, while the origin time is uniformly
distributed. The sampling probability is constant node (square) and
equal to one for the tree in <a href="#fig_example_tree"></a> 
and for the analysis in the exercise given in this tutorial. 
This model represents the phylogenetic continuous-time Markov
chain that links the tree model to the other model components and the
observed sequence data. For more information on probabilistic graphical
models and their notation, please see <a class="citation" href="#Hoehna2014b">(Höhna et al. 2014)</a>.</figcaption>
</figure>

<p>For a more comprehensive description of these models check out the tutorials written by Tracy Heath, April Wright and Walker Pett.</p>

<ul>
  <li><a href="/tutorials/fbd/">Combined-Evidence Analysis and the Fossilized Birth-Death Process for Stratigraphic Range Data</a></li>
  <li><a href="/tutorials/fbd/fbd_specimen">Combined-Evidence Analysis and the Fossilized Birth-Death Process for Analysis of Fossil and Extant Specimens</a></li>
</ul>

<h3 id="the-data">The data</h3>

<p>The sequence data used in this exercise remains the same as the previous exercises (<strong>bears_cytb.nex</strong>). We will also use the same substitution and clock models (the GTR + $\Gamma$ model and the uncorrelated exponential relaxed clock model).</p>

<p>In addition, we’ll take advantage of all the stratigraphic age information available for 20 species in <strong>bears_taxa.tsv</strong>, where <code class="language-plaintext highlighter-rouge">max</code> records the age of the first appearance (i.e. the oldest) and <code class="language-plaintext highlighter-rouge">min</code> records the age of the last appearance (i.e. the youngest) of the species. Time 0.0 represents the present and means the species is extant. We will use the fossilized birth-death process as our tree model, which allows us to directly incorporate fossils into the tree. Note that in this exercise, we will not be including any character for the 12 fossil taxa. This means that we cannot resolve the relationships of these taxa, however, the fossil sampling times are still informative about the FBD model parameters and divergence times.</p>

<p>There’s quite a few additional steps needed to set up this tree model.</p>

<h3 id="the-master-rev-script">The master Rev script</h3>

<p>This time we will start by making some changes to the master Rev script.</p>

<blockquote class="instruction">
  <p>Copy the master script from the previous exercise and call it <strong>MCMC_dating_ex4.Rev</strong>.</p>
</blockquote>

<p>First, at the beginning of you script add the command to read the full list of taxon names from the <strong>bears_taxa.tsv</strong> file using the <code class="language-plaintext highlighter-rouge">readTaxonData</code> function.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>taxa &lt;- readTaxonData("data/bears_taxa.tsv")
</code></pre></div></div>
<p>This function reads a tab-delimited file and creates a variable called <code class="language-plaintext highlighter-rouge">taxa</code> that is a list of all of the taxon names relevant to this analysis, including the minimum/maximum ages.</p>

<p>Delete the helper variable <code class="language-plaintext highlighter-rouge">taxa</code> that we created earlier from the alignment  using the command <code class="language-plaintext highlighter-rouge">taxa &lt;- cytb.taxa()</code>.</p>

<h4 id="add-missing-taxa">Add missing taxa</h4>

<p>Remember that we only have molecular sequence data for the living species. Thus, we must add any taxa that are not found in the molecular (<code class="language-plaintext highlighter-rouge">cytb</code>) partition (i.e. are only found in the fossil data) to that data matrix as missing data (with “?” in place of all characters). In order for all the taxa to appear on the same tree, they all need to be part of the same dataset, as opposed to present in separate datasets. This ensures that there is a unified taxon set that contains all of our tips. We can do this using the <code class="language-plaintext highlighter-rouge">cytb.addMissingTaxa</code> function.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cytb.addMissingTaxa( taxa )
</code></pre></div></div>

<h3 id="the-tree-model">The tree model</h3>

<blockquote class="instruction">
  <p>Start by creating a copy of your <strong>tree_BD.Rev</strong> script, call it <strong>tree_FBD.Rev</strong> and open it in your text editor.</p>
</blockquote>

<h4 id="the-fossilized-birth-death-process">The fossilized birth-death process</h4>

<p>The FBD model has some additional parameters we need to specify. In addition to speciation ($\lambda$), extinction ($\mu$) and extant species sampling ($\rho$), we need to specify the fossil recovery rate ($\psi$), plus the moves on this parameter. Include the following commands anywhere prior to specifying the <code class="language-plaintext highlighter-rouge">timetree</code> variable.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>psi ~ dnExponential(10) 

moves.append( mvScale(psi, lambda=0.5, tune=true, weight=3) )
</code></pre></div></div>
<p>Next delete the command used to specify the root age <code class="language-plaintext highlighter-rouge">extant_mrca</code>.</p>

<p>We will condition the FBD process on the origin time ($\phi$), and specify a uniform distribution on this parameter, with a minimum equal to the oldest first appearance of bears (= 37.2 Ma) and a maximum equal to a prior estimate for the age of caniforms (= 49.0 Ma) that we used in the previous exercise.</p>

<p>Create a stochastic node for the origin time parameter.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>origin_time ~ dnUnif(37.2, 49.0)
</code></pre></div></div>
<p>For this parameter, we will use a sliding window move (<code class="language-plaintext highlighter-rouge">mvSlide</code>). A sliding window samples a parameter uniformly within an interval (defined by the half-width delta). Sliding window moves can be tricky for small values, as the window may overlap zero. However, for parameters such as the origin age, there is little risk of this being an issue.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves.append( mvSlide(origin_time, delta=1.0, tune=true, weight=5.0) )
</code></pre></div></div>
<p>Note that the biological interpretation of this parameter is not that straightforward, so we can think of this as a nuisance parameter.</p>

<p>Next, we need to change the tree model distribution from the birth-death process to the FBD range process using the <code class="language-plaintext highlighter-rouge">dnFBDRP</code> command. Note that this function takes a different set of arguments to <code class="language-plaintext highlighter-rouge">dnBDP</code>.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree_dist = dnFBDRP(lambda=speciation_rate, mu=extinction_rate, psi=psi, rho=rho, origin=origin_time, taxa=taxa)
</code></pre></div></div>

<h4 id="clade-constraints">Clade constraints</h4>

<p>Again we’ll constrain the clade Ursinae to be monophyletic, but this time we have one fossil species to add this constraint: <em>Ursus abstrusus</em>. Add this taxon to your clade constraints before specifying the stochastic <code class="language-plaintext highlighter-rouge">timetree</code> node.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>clade_ursinae = clade("Melursus_ursinus", "Ursus_arctos", "Ursus_maritimus", 
                  "Helarctos_malayanus", "Ursus_americanus", "Ursus_thibetanus", "Ursus_abstrusus") 
constraints = v(clade_ursinae)
</code></pre></div></div>

<h4 id="moves-on-the-tree">Moves on the tree</h4>

<p>In addition to the move we applied previously to sample the tree topology, <code class="language-plaintext highlighter-rouge">mvFNPR</code>, we also need to add an extra move <code class="language-plaintext highlighter-rouge">mvCollapseExpandFossilBranch</code> that will change a fossil that is a sampled ancestor so that it is on its own branch and vice versa.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves.append( mvCollapseExpandFossilBranch(timetree, origin_time, weight=6.0) )
</code></pre></div></div>
<p>In addition, when conditioning on the origin time, we also need to explicitly sample the root age (<code class="language-plaintext highlighter-rouge">mvRootTimeSlideUniform</code>).</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves.append( mvRootTimeSlideUniform(timetree, origin_time, weight=5.0) )
</code></pre></div></div>

<h4 id="monitoring-parameters-of-interest-using-deterministic-nodes">Monitoring parameters of interest using deterministic nodes</h4>

<p>We are still interested in the age of the MRCA of all living bears (<code class="language-plaintext highlighter-rouge">extant_mrca</code>). However, since now we have extinct species in our tree, the root of the tree may not represent the MRCA of all extant species. To monitor the age of this node in our MCMC sample, we must use the <code class="language-plaintext highlighter-rouge">clade</code> function to define the node. Importantly, since we did not include this clade in our constraints that defined <code class="language-plaintext highlighter-rouge">timetree</code>, this clade will not be constrained to be monophyletic. Once this clade is defined we can instantiate a deterministic node called <code class="language-plaintext highlighter-rouge">extant_mrca</code> with the <code class="language-plaintext highlighter-rouge">tmrca</code> function that will record the age of the MRCA of all living bears.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>clade_extant = clade("Ailuropoda_melanoleuca","Tremarctos_ornatus","Melursus_ursinus",
                    "Ursus_arctos","Ursus_maritimus","Helarctos_malayanus",
                    "Ursus_americanus","Ursus_thibetanus")
extant_mrca := tmrca(timetree, clade_extant)
</code></pre></div></div>
<p>We can also create deterministic node to compute and keep track of the number of sampled ancestors (<code class="language-plaintext highlighter-rouge">num_samp_anc</code>) in our <code class="language-plaintext highlighter-rouge">timetree</code>.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>num_samp_anc := timetree.numSampledAncestors()
</code></pre></div></div>
<p>Finally, we will monitor the tree after removing taxa for which we did not include any character data. The phylogenetic placement of these taxa is based only on their occurrence times and any clade constraints we applied. Because no data are available to resolve their relationships relative to other lineages, we will treat their placement as nuisance parameters and remove them from the output. Create a vector of extinct taxa and use the function <code class="language-plaintext highlighter-rouge">fnPruneTree</code> to create a deterministic node for the tree excluding extinct samples.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>extinct_sp = v("Agriarctos_spp", "Ailurarctos_lufengensis", "Ballusia_elmensis",
				"Indarctos_arctoides", "Indarctos_punjabiensis", "Indarctos_vireti",
				"Kretzoiarctos_beatrix", "Parictis_montanus", "Ursavus_brevirhinus",
				"Ursavus_primaevus", "Ursus_abstrusus", "Zaragocyon_daamsi")

pruned_tree := fnPruneTree(timetree, prune=extinct_sp)
</code></pre></div></div>

<h3 id="back-to-the-master-rev-script">Back to the master Rev script</h3>

<p>As before, change the file used to specify the tree model from <strong>tree_BD_nodedate.Rev</strong> to <strong>tree_FBD.Rev</strong>.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("scripts/tree_FBD.Rev")
</code></pre></div></div>
<p>Second, update the name of the output files. We also need to update the name of the tree we’re printing to file from the <code class="language-plaintext highlighter-rouge">timetree</code> variable that incorporates extinct taxa to the <code class="language-plaintext highlighter-rouge">pruned_tree</code> variable that includes our extant species only.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors.append( mnModel(filename="output/bears_FBD.log", printgen=10) )
monitors.append( mnFile(filename="output/bears_FBD.trees", printgen=10, pruned_tree) )
</code></pre></div></div>
<p>Don’t forget to update the commands used to generate the summary tree.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>trace = readTreeTrace("output/bears_FBD.trees")
mccTree(trace, file="output/bears_FBD.mcc.tre" )
</code></pre></div></div>

<blockquote class="instruction">
  <p>Run your MCMC analysis!</p>
</blockquote>

<h3 id="examining-the-output">Examining the output</h3>

<p>Let’s examine the output in Tracer and compare our results to those generated using node dating.</p>

<blockquote class="instruction">
  <p>Open the program Tracer and load the log files <strong>bears_FBD.log</strong> and <strong>bears_nodedate.log</strong>.</p>
</blockquote>

<p>Compare the node age estimates for the MRCA of extant bears. What do you notice about the different marginal densities?</p>

<figure id="fig_trace"><p><img src="figures/bears_FBD_nodedate.png" width="700" /></p>
<figcaption>The Marginal Density panel in Tracer showing the posterior estimates for the MRCA of extant bears using two different calibration appraoches (node dating versus the FBD model). To reproduce this figure select Colour by: Trace file and Legend: Top-Right.</figcaption>
</figure>

<p>You may also notice that the age recovered for <code class="language-plaintext highlighter-rouge">age_ursinae</code> is also a bit different but we need to be careful - the estimates obtained for this variable between the node dating and FBD analyses are not directly comparable. Recall that we added an additional fossil taxon to the Ursinae constraint in this analysis. What steps would you have taken to recover an equivalent parameter, without eliminating the <code class="language-plaintext highlighter-rouge">clade_ursinae</code> constraint?</p>

<h4 id="the-tree-output">The tree output</h4>

<p>Have a quick look at your MCC tree in FigTree. How does it compare the tree your recovered using node dating?</p>

<h3 id="next">Next</h3>

<p>It’ll be more fun to examine the tree output once we infer the phylognetic position of extinct taxa and sampled ancestors, which we’ll do in the next exercise by incorporating morphological character data.</p>

<blockquote class="instruction">
  <p>Click below to begin the next exercise!</p>
</blockquote>

<ul>
  <li><a href="/tutorials/dating/tefbd">Estimating speciation times using total-evidence dating</a></li>
</ul>

<p>For further options and information about the models used in this exercise see Tracy Heath, April Wright and Walker Pett’s tutorial <a href="/tutorials/fbd/">Combined-Evidence Analysis and the Fossilized Birth-Death Process for Stratigraphic Range Data</a></p>

<ol class="bibliography"><li><span id="Heath2014">Heath T.A., Huelsenbeck J.P., Stadler T. 2014. The fossilized birth-death process for coherent calibration of divergence-time estimates. Proceedings of the National Academy of Sciences. 111:E2957–E2966.</span>

<a href="https://doi.org/10.1073/pnas.1319091111">10.1073/pnas.1319091111</a>

</li>
<li><span id="Hoehna2014b">Höhna S., Heath T.A., Boussau B., Landis M.J., Ronquist F., Huelsenbeck J.P. 2014. Probabilistic Graphical Model Representation in Phylogenetics. Systematic Biology. 63:753–771.</span>

<a href="https://doi.org/10.1093/sysbio/syu039">10.1093/sysbio/syu039</a>

</li>
<li><span id="Stadler2010">Stadler T. 2010. Sampling-through-time in birth-death trees. Journal of Theoretical Biology. 267:396–404.</span>

<a href="https://doi.org/10.1016/j.jtbi.2010.09.010">10.1016/j.jtbi.2010.09.010</a>

</li>
<li><span id="Stadler2018">Stadler T., Gavryushkina A., Warnock R.C.M., Drummond A.J., Heath T.A. 2018. The fossilized birth-death model for the analysis of stratigraphic range data under different speciation modes. Journal of Theoretical Biology. 447:41–55.</span>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-0Y03X9Q2TJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-0Y03X9Q2TJ');
</script>

  </body>
</html>
